<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>1. From geospatial to spatial transcriptomics • SFEWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="1. From geospatial to spatial transcriptomics">
<meta property="og:description" content="SFEWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SFEWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vig1_intro.html">1. From geospatial to spatial transcriptomics</a>
    </li>
    <li>
      <a href="../articles/vig2_SFE.html">2. SpatialFeatureExperiment</a>
    </li>
    <li>
      <a href="../articles/vig3_Voyager.html">3. Voyager</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lambdamoses/SFEWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>1. From geospatial to spatial
transcriptomics</h1>
                        <h4 data-toc-skip class="author">Lambda
Moses</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:dlu2@caltech.edu" class="email">dlu2@caltech.edu</a>
      
                              <h4 data-toc-skip class="author">Lior
Pachter</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:lpachter@caltech.edu" class="email">lpachter@caltech.edu</a>
      
                  
            <h4 data-toc-skip class="date">r Sys.Date()</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/lambdamoses/SFEWorkshop/blob/HEAD/vignettes/vig1_intro.Rmd"><code>vignettes/vig1_intro.Rmd</code></a></small>
      <div class="hidden name"><code>vig1_intro.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="an-analogy-from-geographical-space">An analogy from geographical space<a class="anchor" aria-label="anchor" href="#an-analogy-from-geographical-space"></a>
</h2>
<p>Data from bulk RNA-seq, single cell RNA-seq (scRNA-seq), and spatial
transcriptomics all come from originally spatially organized tissues
(except for blood), where such spatial organization plays a role in the
properties and functions of the tissues. For this reason, a geographical
analogy may be more adequate than the traditional analogy of the
smoothie vs. fruit salad vs. tart.</p>
<p>Many of you attending this conference in person have traveled to
Seattle, and might explore Seattle after the talks or the entire
conference. Bulk RNA-seq is like average values over the entire
Washington state. It gives some information, such as that Washington
tends to be cooler and more humid than California, and that median
household income in California and Washington are higher than that in
Nevada and Oregon. However, a tourist would like a higher
resolution.</p>
<p><img src="vig1_intro_files/figure-html/unnamed-chunk-6-1.png" width="1200" style="display: block; margin: auto;"></p>
<p>Non-spatial scRNA-seq is like tourist attractions in Seattle
separated from their neighborhoods. There’s the Space Needle, the Museum
of Pop Culture, Chihuli glass museum, Smith Tower, Pike Place, Museum of
Flight, and etc. These tourist attractions can also be classified based
on their characteristics, such as architectural style, function, and era
of construction. Dimension reduction can be performed on these numerous
characteristics. This would be of more interest to an avid tourist who
really wants to get to know Seattle than the state level averages.</p>
<p><img src="seattle_attractions.png" width="256" style="display: block; margin: auto;"></p>
<p>How to get to these tourist attractions? When you look at the map,
you find that certain kinds of tourist attractions tend to cluster in
space, such as the cluster of museums in the vicinity of Space Needle
and the cluster of older buildings around Smith Tower. Different regions
of Seattle with different vibes and functions are also annotated on the
map. There are historical reasons that led to such spatial regions and
clustering, such as the 1962 World’s Fair that gave rise to the Space
Needle. Here we see how locating the tourist attractions in space point
to a deeper understanding of the properties of Seattle. Spatial
transcriptomics is like studying a map of Seattle.</p>
<p><img src="seattle_map.png" width="256" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="introduction-to-spatial-transcriptomics-data">Introduction to spatial transcriptomics data<a class="anchor" aria-label="anchor" href="#introduction-to-spatial-transcriptomics-data"></a>
</h2>
Back to spatial transcriptomics. A working definition of “spatial
transcriptomics” is any attempt to quantify the transcripts of more
genes than can be done in one round of fluorescent <em>in situ</em>
hybridization (FISH) while preserving the spatial context of the gene
expression in the tissue. Since 2018, interest in spatial
transcriptomics – in both data collection and data analysis – has grown
drastically.
<div class="figure" style="text-align: center">
<img src="https://pachterlab.github.io/LP_2021/07-current_analysis_files/figure-html/analysis-current-1.png" alt="Number of publication over time for current era data collection and data analysis. Bin width is 120 days. The x-shaped points show the number of publications from the last bin, which is not yet full."><p class="caption">
Number of publication over time for current era data collection and data
analysis. Bin width is 120 days. The x-shaped points show the number of
publications from the last bin, which is not yet full.
</p>
</div>
<p>Based on how the spatial context is preserved, spatial
transcriptomics data collection technologies fall into 5 categories,
sorted by popularity (plot showing number of publications per category),
though there are gray areas:</p>
<div class="figure" style="text-align: center">
<img src="https://pachterlab.github.io/LP_2021/04-current_files/figure-html/n-insts-1.png" alt="Techniques used by at least 3 institutions and the number of institutions that have used them."><p class="caption">
Techniques used by at least 3 institutions and the number of
institutions that have used them.
</p>
</div>
<ol style="list-style-type: decimal">
<li>Next generation sequencing (NGS) with spatial barcoding, where
spatial locations of the barcodes are known, such as 10X Visium,
Slide-seq, and Stereo-seq. Many of these methods, such as the popular
Visium, do not have single cell resolution. Hence cell type
deconvolution is commonly performed.</li>
<li>Region of interest (ROI) selection, where the act of specifying the
ROIs itself defines the spatial locations, such as laser capture
microdissection (LCM) followed by NGS and GeoMX DSP, where UV light is
used to cleave probes only in the selected ROIs for quantification. The
ROIs can also be sections in one spatial dimension, such as in Tomo-seq.
These methods also generally don’t have single cell resolution.</li>
<li>Single molecular FISH, where each detected transcript can be seen as
a discrete puncta and counted, such as MERFISH, seqFISH, and Molecular
Cartography. Combinatorial barcoding over multiple rounds – where
fluorescent probes are hybridized to the transcripts or cDNAs and color
or presence or absence of the probes are recorded and genes are identify
by the sequence of colors or presence/absence – is used to quantify more
genes than easily discernible colors simultaneously. Datasets from
highly multiplexed smFISH are usually not genome wide, and only quantify
a few hundred genes at a time, though these technologies have been used
in an increasing number of cells per experiment.</li>
<li>
<em>In situ</em> sequencing (ISS), which is similar to smFISH as the
transcripts are visualized as discrete puncta and counted. However, in
ISS, the gene barcodes are determined by <em>in situ</em> sequencing,
usually with sequencing by ligation, rather than with combinatorial
barcoding. smFISH and ISS based data have single cell resolution.</li>
<li>Spatial context is not directly recorded, but computationally
reconstructed, such as by recording pairwise spatial adjacency of cDNAs
as in DNA microscopy. In addition, the term “spatial” is sometimes used
to refer to sub-cellular resolution, such as in APEX-seq, where
transcripts are “localized” or organelles; since no spatial coordinates
are recorded, packages in this workshop are irrelevant to these
technologies.</li>
</ol>
<p><img src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs41592-022-01409-2/MediaObjects/41592_2022_1409_Fig2_HTML.png" style="display: block; margin: auto;"></p>
<p>Data analysis methods written for spatial transcriptomics can be
broadly categorized as upstream and downtream. In upstream analysis, the
raw data is converted into more usable forms, such as getting the gene
count matrix from fastq files and cell type deconvolution of Visium
spots. Downstream analysis begins with the more usable form of data for
further biological inferences, such as finding spatially variable genes,
spatial regions informed by gene expression, and cell-cell interactions.
<code>SpatialFeatureExperiment</code>, as a way to represent data, is
more upstream, while <code>Voyager</code>, for exploratory spatial data
analysis (ESDA), is a little more downstream.</p>
<p>In the literature when spatial transcriptomics data is generated, the
data is often treated as non-spatial scRNA-seq data, and some data
analysis methods for spatial transcriptomics data do not take the
spatial information into account and are sometimes aimed at non-spatial
data as well. For example, most cell type deconvolution methods don’t
take into account spatial autocorrelation in cell type distribution, and
deconvolution methods for bulk RNA-seq is sometimes used for Visium
data. However, the spatial information presents many opportunities
unavailable to scRNA-seq, and this workshop will explore some of these
opportunities. Among the opporunities are:</p>
<ol style="list-style-type: decimal">
<li>Identifying spatial regions characterized by gene expression of cell
types, just like neighborhoods in geographical space. Many methods have
been written for this purpose.</li>
<li>Tissue and cellular morphology is available along side
transcriptomics, and may be related to gene expression. This is explored
a little in this workshop.</li>
<li>Explore how gene expression and quality control (QC) metrics vary in
space, such as strength, length scale, and direction (anisotropy) of
spatial autocorrelation. This workshop explores strength and length
scale of spatial autocorrelation; anisotropy will be supported in a
later version of the <code>Voyager</code> package.</li>
</ol>
</div>
<div class="section level2">
<h2 id="geospatial-data-types-and-how-they-may-relate-to-spatial-transcriptomics">Geospatial data types and how they may relate to spatial
transcriptomics<a class="anchor" aria-label="anchor" href="#geospatial-data-types-and-how-they-may-relate-to-spatial-transcriptomics"></a>
</h2>
<p>Geospatial data broadly fall into two types by representation: vector
and raster. Vector data represents the world as points, lines, and
polygons, specified by coordinates. For example, polygons would be
specified by coordinates of the vertices. Raster data are basically
images, where each pixel has a value, though unlike the typical RGB
image, raster data can have many different layers analogous to the 3
channels in RGB images. Raster is common in remote sensing.
<img src="spatial_types.png" width="319" style="display: block; margin: auto;"></p>
<p>Vector data can be further classified by the processes generating the
data:</p>
<ol style="list-style-type: decimal">
<li>Areal, where values are for entire areas (e.g. cities, zip codes)
rather than individuals. This is common in epidemiology to protect
patients’ privacy.</li>
<li>Geostatistical, where a spatial process (e.g. Gaussian process) is
sampled at points in space, and the locations of those points are known
and not of interest by themselves. Rather, the values are of interest.
An example is measurements of pollutant concentration at points in
space.</li>
<li>Point process, where the locations of occurrence is of interest,
such as where individuals of a species of plants occur. The locations
themselves are analyzed to find if they are randomly distributed,
clustered, or repelling each other, and if they are affected by
covariates.</li>
<li>When a phenomenon can only occur on a spatial network, such as roads
and rivers. This can be a point process, such as locations of traffic
incidents. This can also be analogous to raster data, such as Strava’s
road/trail popularity heatmap, or areal when Strava’s heatmap is
summarized over segments.</li>
</ol>
<p>For each of these data types, there are already well-established
tools for data analysis. For example, <code>sf</code> represents vector
data and makes it behave like a data frame in R, <code>spdep</code> can
be used for spatial dependency of areal data, <code>INLA</code> can be
used for Bayesian spatial modeling, <code>gstat</code> is used for
interpolation with geostatistical data (kriging), <code>spatstat</code>
is used for point process data, <code>sfnetworks</code> for network
data, and <code>spatstat</code> supports point process analysis within a
network.</p>
<p>If we can relate these data types to spatial transcriptomics, then we
can take advantage of these data analysis tools:</p>
<ol style="list-style-type: decimal">
<li>Visium, LCM, and GeoMX DSP data aggregate transcript counts over
areas. In addition, in smFISH and ISS data, the counts are also often
aggregated over cells, and the cells are represented as polygons in
space. Here such data is analogous the areal data, although unlike for
cities and zip codes, the polygons may not be contiguous most of the
time.</li>
<li>If gene expression, or gene expression program, are imaged as
spatial processes, and the cells or Visium spots are imagined as points,
ignoring their area, then the data is analogous to geostatistical data.
(Shall I mention that R package using kriging)</li>
<li>Locations of transcript spots in smFISH and ISS based data, and
locations of cells and cell types can be thought of as spatial point
processes.</li>
<li>Spatial networks seem less relevant to spatial transcriptomics. One
potential use case is blood vessels. However, since spatial
transcriptomics data is usually obtained from thin sections of the
original 3D tissue, very little of the original 3D blood vessel network
remains in the section.</li>
<li>With the regular grids, Visium and 2016 Spatial Transcriptomics data
may be seen as a low resolution raster.</li>
</ol>
<p>Some geospatial data analysis methods have already been used in
spatial transcriptomics, such as Gaussian process regression (which
kriging is based on), Potts model, spatial point process analyses,
conditional autoregressive models, and global and local variants of
Moran’s I. (get the references) However, these sptial transcriptomics
packages and data analysis in the literature don’t fully explore the
opportunities given by the spatial information mentioned above, nor do
they always take advantage of the existing software infrastructure
originally built for geospatial data.</p>
<p>This workshop presents the packages
<code>SpatialFeatureExperiment</code> and <code>Voyager</code>, which
bring more of the existing vector geospatial data analysis tools to
spatial transcriptomics. <code>SpatialFeatureExperiment</code> (SFE)
implements a new S4 class extending <code>SpatialExperiment</code> with
<code>sf</code> data frames representing cell or Visium spot polygons,
geometries of other objects, anatomical regions, and tissue boundaries,
and spatial neighborhood graphs. <code>Voyager</code> to SFE is just
like <code>scater</code>, <code>scran</code>, and <code>scuttle</code>
to <code>SingleCellExperiment</code>, implementing basic ESDA using
well-established geospatial tools and plotting functions for geometries
and ESDA results. We focus on vector data here. Although the fluorsecent
or H&amp;E images are raster, features extracted from the images, such
as cell and nuclei segmentation, and transcript spots, are more commonly
used for downstream analyses.</p>
</div>
<div class="section level2">
<h2 id="limitations-and-future-directions">Limitations and future directions<a class="anchor" aria-label="anchor" href="#limitations-and-future-directions"></a>
</h2>
<p>No analogy is perfect, and these are some limitations of the
geospatial analogy:</p>
<ol style="list-style-type: decimal">
<li>Biological tissues are usually originally 3D. Although most spatial
transcriptomics datasets come from thin sections, some datasets come
from thicker sections. However, the z dimension usually has much lower
resolution than x and y if present, and might need to be treated
differently. Geospatial tools and resources are very focused on 2D. If
3D support is present (e.g. 3D spatial point processes in
<code>spatstat</code>), the support may be limited.</li>
<li>Spatial transcriptomics datasets can get much larger than most
geospatial vector datasets, from thousands of genes and thousands to
millions of cells or spots. There should be more emphasis on
multivariate spatial analyses, which we plan to add to a later version
of <code>Voyager</code>. On disk tools exist for geospatial data, but
are not supported by SFE yet.</li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Lambda Moses, Lior Pachter.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
