<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2. SpatialFeatureExperiment • SFEWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2. SpatialFeatureExperiment">
<meta property="og:description" content="SFEWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SFEWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vig1_intro.html">1. From geospatial to spatial transcriptomics</a>
    </li>
    <li>
      <a href="../articles/vig2_SFE.html">2. SpatialFeatureExperiment</a>
    </li>
    <li>
      <a href="../articles/vig3_Voyager.html">3. Voyager</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lambdamoses/SFEWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2. SpatialFeatureExperiment</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/lambdamoses/SFEWorkshop/blob/HEAD/vignettes/vig2_SFE.Rmd"><code>vignettes/vig2_SFE.Rmd</code></a></small>
      <div class="hidden name"><code>vig2_SFE.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="class-structure">Class structure<a class="anchor" aria-label="anchor" href="#class-structure"></a>
</h2>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p>Simple Features are a standardized way to represent geometries in
space. The <a href="https://cran.r-project.org/web/packages/sf/index.html" class="external-link"><code>sf</code></a>
package provides Simple Features access to R, with the <code>sf</code>
data frame, which can behave like a regular data frame but has a special
list column for the geometries.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"https://user-images.githubusercontent.com/520851/50280460-e35c1880-044c-11e9-9ed7-cc46754e49db.jpg"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="https://user-images.githubusercontent.com/520851/50280460-e35c1880-044c-11e9-9ed7-cc46754e49db.jpg" alt=""><p class="caption">Schematic of the sf data frame, illustration by
Allison Horst</p>
</div>
<p><code>SpatialFeatureExperiment</code> (SFE) is a new <a href="http://adv-r.had.co.nz/S4.html" class="external-link">S4</a> class built on top of <a href="https://bioconductor.org/packages/release/bioc/html/SpatialExperiment.html" class="external-link"><code>SpatialExperiment</code></a>
(SPE). <code>SpatialFeatureExperiment</code> incorporates geometries and
geometry operations with the <code>sf</code> package. Examples of
supported geometries are Visium spots represented with polygons
corresponding to their size, cell or nuclei segmentation polygons,
tissue boundary polygons, pathologist annotation of histological
regions, and transcript spots of genes. Using <code>sf</code>,
<code>SpatialFeatureExperiment</code> leverages the GEOS C++ libraries
underlying <code>sf</code> for geometry operations, including algorithms
for for determining whether geometries intersect, finding intersection
geometries, buffering geometries with margins, etc. A schematic of the
SFE object is shown below:</p>
<div class="figure">
<img src="sfe_schematics.png" alt="SpatialFeatureExperiment expands on SpatialExperiment by adding column, row, and annotation geometries and spatial graphs. This is explained in detail in the following paragraphs." width="100%"><p class="caption">
Schematics of the SFE object
</p>
</div>
<p>Below is a list of SFE features that extend the SPE object:</p>
<ul>
<li>
<code>colGeometries</code> are <code>sf</code> data frames
associated with the entities that correspond to columns of the gene
count matrix, such as Visium spots or cells. The geometries in the
<code>sf</code> data frames can be Visium spot centroids, Visium spot
polygons, or for datasets with single cell resolution, cell or nuclei
segmentations. Multiple <code>colGeometries</code> can be stored in the
same SFE object, such as one for cell segmentation and another for
nuclei segmentation. There can be non-spatial, attribute columns in a
<code>colGeometry</code> rather than <code>colData</code>, because the
<code>sf</code> class allows users to specify how attributes relate to
geometries, such as “constant”, “aggregate”, and “identity”. See the
<code>agr</code> argument of the <a href="https://r-spatial.github.io/sf/reference/sf.html" class="external-link"><code>st_sf</code>
documentation</a>.</li>
<li>
<code>colGraphs</code> are spatial neighborhood graphs of cells or
spots. The graphs have class <code>listw</code> (<code>spdep</code>
package), and the <code>colPairs</code> of
<code>SingleCellExperiment</code> was not used so no conversion is
necessary to use the numerous spatial dependency functions from
<code>spdep</code>, such as those for Moran’s I, Geary’s C, Getis-Ord
Gi*, LOSH, etc. Conversion is also not needed for other classical
spatial statistics packages such as <code>spatialreg</code> and
<code>adespatial</code>.</li>
<li>
<code>rowGeometries</code> are similar to
<code>colGeometries</code>, but support entities that correspond to rows
of the gene count matrix, such as genes. A potential use case is to
store transcript spots for each gene in smFISH or in situ sequencing
based datasets.</li>
<li>
<code>rowGraphs</code> are similar to <code>colGraphs</code>. A
potential use case may be spatial colocalization of transcripts of
different genes.</li>
<li>
<code>annotGeometries</code> are <code>sf</code> data frames
associated with the dataset but not directly with the gene count matrix,
such as tissue boundaries, histological regions, cell or nuclei
segmentation in Visium datasets, and etc. These geometries are stored in
this object to facilitate plotting and using <code>sf</code> for
operations such as to find the number of nuclei in each Visium spot and
which histological regions each Visium spot intersects. Unlike
<code>colGeometries</code> and <code>rowGeometries</code>, the number of
rows in the <code>sf</code> data frames in <code>annotGeometries</code>
is not constrained by the dimension of the gene count matrix and can be
arbitrary.</li>
<li>
<code>annotGraphs</code> are similar to <code>colGraphs</code> and
<code>rowGraphs</code>, but are for entities not directly associated
with the gene count matrix, such as spatial neighborhood graphs for
nuclei in Visium datasets, or other objects like myofibers. These graphs
are relevant to <code>spdep</code> analyses of attributes of these
geometries such as spatial autocorrelation in morphological metrics of
myofibers and nuclei. With geometry operations with <code>sf</code>,
these attributes and results of analyses of these attributes
(e.g. spatial regions defined by the attributes) may be related back to
gene expression.</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/dev/" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span></span>
<span><span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span></span>
<span><span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span><span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span></span>
<span><span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SFEData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1; sf_use_s2() is TRUE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Matrix'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example dataset</span></span>
<span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-07-15</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 77 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(77): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCTTTGGTCGCGACG</span></span>
<span><span class="co">#&gt;   TTGATGTGTAGTCCCG</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="geometries">Geometries<a class="anchor" aria-label="anchor" href="#geometries"></a>
</h3>
<p>User interfaces to get or set the geometries and spatial graphs
emulate those of <a href="https://rdrr.io/bioc/SingleCellExperiment/man/reducedDims.html" class="external-link"><code>reducedDims</code>
in <code>SingleCellExperiment</code></a>.</p>
<p>Geometries associated with columns of the gene count matrix (cells or
Visium spots) can be get/set with <code>colGeometry</code>. Similarly,
geometries associated with rows of the gene count matrix (genes or
features) can be get/set with <code>rowGeometry</code>, and geometries
associated with the annotations can be get/set with
<code>annotGeometry</code>. Here we demonstrate <code>colGeometry</code>
as the arguments for <code>rowGeometry</code> and
<code>annotGeometry</code> are the same. See the <a href="https://pachterlab.github.io/SpatialFeatureExperiment/dev/articles/SFE.html" class="external-link">vignette
of <code>SpatialFeatureExperiment</code></a> for more details.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get Visium spot polygons</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 77 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5000 ymin: 13000 xmax: 7000 ymax: 15000</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                                        geometry sample_id</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG POLYGON ((6472.186 13875.23...     Vis5A</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG POLYGON ((5778.291 13635.43...     Vis5A</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT POLYGON ((7000 13809.84, 69...     Vis5A</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT POLYGON ((6749.535 13874.64...     Vis5A</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC POLYGON ((5500.941 13636.03...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT POLYGON ((6612.42 14598.82,...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGCGACTCCTG POLYGON ((5501.981 14118.62...     Vis5A</span></span>
<span><span class="co">#&gt; AATTCATAAGGGATCT POLYGON ((6889.769 14598.22...     Vis5A</span></span>
<span><span class="co">#&gt; ACGAGTACGGATGCCC POLYGON ((5084.397 13395.63...     Vis5A</span></span>
<span><span class="co">#&gt; ACGCTAGTGATACACT POLYGON ((5639.096 13394.44...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use st_geometry so it won't color by the attributes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="va">spots</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_SFE_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"foobar"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
<p><code>colGeometryNames</code> gets or sets the names of the
geometries</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">cg_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">colGeometryNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "spotPoly" "foobar"</span></span></code></pre></div>
<p>There are shorthands for some specific geometries. For example,
<code>spotPoly(sfe)</code> is equivalent to
<code>colGeometry(sfe, "spotPoly")</code> for Visium spot polygons, and
<code>txSpots(sfe)</code> is equivalent to
<code>rowGeometry(sfe, "txSpots")</code> for transcript spots in single
molecule technologies.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">spots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 77 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5000 ymin: 13000 xmax: 7000 ymax: 15000</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                                        geometry sample_id</span></span>
<span><span class="co">#&gt; AAATTACCTATCGATG POLYGON ((6472.186 13875.23...     Vis5A</span></span>
<span><span class="co">#&gt; AACATATCAACTGGTG POLYGON ((5778.291 13635.43...     Vis5A</span></span>
<span><span class="co">#&gt; AAGATTGGCGGAACGT POLYGON ((7000 13809.84, 69...     Vis5A</span></span>
<span><span class="co">#&gt; AAGGGACAGATTCTGT POLYGON ((6749.535 13874.64...     Vis5A</span></span>
<span><span class="co">#&gt; AATATCGAGGGTTCTC POLYGON ((5500.941 13636.03...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGATACGCTAT POLYGON ((6612.42 14598.82,...     Vis5A</span></span>
<span><span class="co">#&gt; AATGATGCGACTCCTG POLYGON ((5501.981 14118.62...     Vis5A</span></span>
<span><span class="co">#&gt; AATTCATAAGGGATCT POLYGON ((6889.769 14598.22...     Vis5A</span></span>
<span><span class="co">#&gt; ACGAGTACGGATGCCC POLYGON ((5084.397 13395.63...     Vis5A</span></span>
<span><span class="co">#&gt; ACGCTAGTGATACACT POLYGON ((5639.096 13394.44...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">spots</span></span></code></pre></div>
<p><code>tissueBoundary(sfe)</code> is equivalent to
<code>annotGeometry(sfe, "tissueBoundary")</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Getter</span></span>
<span><span class="op">(</span><span class="va">tb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 1 feature and 2 fields</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 5094 ymin: 13000 xmax: 7000 ymax: 14969</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt;   ID                       geometry sample_id</span></span>
<span><span class="co">#&gt; 7  7 POLYGON ((5094 13000, 5095 ...     Vis5A</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="va">tb</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_SFE_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Setter</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tb</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="spatial-graphs">Spatial graphs<a class="anchor" aria-label="anchor" href="#spatial-graphs"></a>
</h3>
<p>Spatial dependence analyses in the <code>spdep</code> package
requires a spatial neighborhood graph. The
<code>SpatialFeatureExperiment</code> package wraps functions in the
<code>spdep</code> package to find spatial neighborhood graphs. In this
example, triangulation is used to find the spatial graph; many other
methods are also supported, such as k nearest neighbors, distance based
neighbors, and polygon contiguity.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe</span>, MARGIN <span class="op">=</span> <span class="fl">2</span>, type <span class="op">=</span> <span class="st">"spotPoly"</span>, method <span class="op">=</span> <span class="st">"tri2nb"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 428 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.218755 </span></span>
<span><span class="co">#&gt; Average number of links: 5.558442 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 28.24222 310.1917</span></span></code></pre></div>
<p>This is what the triangulated graph looks like</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">g</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_SFE_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Similar to <code>colGeometry</code>, <code>rowGeometry</code>, and
<code>annotGeometry</code>, the spatial graphs can be get and set with
<code>colGraph</code>, <code>rowGraph</code>, and
<code>annotGraph</code>; the arguments of the graph getters and setters
are the same as those of the geometry getters and setters, so we only
demonstrate <code>colGraph</code> here.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set graph by name</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, type <span class="op">=</span> <span class="st">"graph1"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">g</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get graph by name</span></span>
<span><span class="op">(</span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"graph1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Characteristics of weights list object:</span></span>
<span><span class="co">#&gt; Neighbour list object:</span></span>
<span><span class="co">#&gt; Number of regions: 77 </span></span>
<span><span class="co">#&gt; Number of nonzero links: 428 </span></span>
<span><span class="co">#&gt; Percentage nonzero weights: 7.218755 </span></span>
<span><span class="co">#&gt; Average number of links: 5.558442 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Weights style: W </span></span>
<span><span class="co">#&gt; Weights constants summary:</span></span>
<span><span class="co">#&gt;    n   nn S0       S1       S2</span></span>
<span><span class="co">#&gt; W 77 5929 77 28.24222 310.1917</span></span></code></pre></div>
<p>For Visium, spatial neighborhood graph of the hexagonal grid can be
found with the known locations of the barcodes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span></code></pre></div>
<p>This is what the Visium graph looks like</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"visium"</span><span class="op">)</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_SFE_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>Similar to the geometries, the graphs have getters and setters for
the names:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">colGraphNames</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "graph1" "visium"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="multiple-samples">Multiple samples<a class="anchor" aria-label="anchor" href="#multiple-samples"></a>
</h3>
<p>Thus far, the example dataset used only has one sample. The
<code>SpatialExperiment</code> (SPE) object has a special column in
<code>colData</code> called <code>sample_id</code>, so data from
multiple tissue sections can coexist in the same SPE object for joint
dimension reduction and clustering while keeping the spatial coordinates
separate. It’s important to keep spatial coordinates of different tissue
sections separate because first, the coordinates would only make sense
within the same section, and second, the coordinates from different
sections can have overlapping numeric values.</p>
<p>SFE inherits from SPE, and with geometries and spatial graphs,
<code>sample_id</code> is even more important. The geometry and graph
getter and setter functions shown above have a <code>sample_id</code>
argument, which is optional when only one sample is present in the SFE
object. This argument is mandatory if multiple samples are present, and
can be a character vector for multiple samples or “all” for all samples.
Below are examples of using the getters and setters for multiple
samples.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Construct toy dataset with 2 samples</span></span>
<span><span class="va">sfe1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small"</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-07-15</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">sfe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"small2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-07-15</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe2</span><span class="op">)</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="st">"sample02"</span></span>
<span><span class="op">(</span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sfe1</span>, <span class="va">sfe2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 149 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(149): AAATTACCTATCGATG AACATATCAACTGGTG ... TTCCTCGGACTAACCA</span></span>
<span><span class="co">#&gt;   TTCTGACCGGGCTCAA</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (GEOMETRY), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A: </span></span>
<span><span class="co">#&gt; sample02:</span></span></code></pre></div>
<p>Use the <code>sampleIDs</code> function to see the names of all
samples</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/sampleIDs.html" class="external-link">sampleIDs</a></span><span class="op">(</span><span class="va">sfe_combined</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Vis5A"    "sample02"</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only get the geometries for the second sample</span></span>
<span><span class="op">(</span><span class="va">spots2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">colGeometry</a></span><span class="op">(</span><span class="va">sfe_combined</span>, <span class="st">"spotPoly"</span>, sample_id <span class="op">=</span> <span class="st">"sample02"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Simple feature collection with 72 features and 1 field</span></span>
<span><span class="co">#&gt; Geometry type: POLYGON</span></span>
<span><span class="co">#&gt; Dimension:     XY</span></span>
<span><span class="co">#&gt; Bounding box:  xmin: 6000 ymin: 7025.865 xmax: 8000 ymax: 9000</span></span>
<span><span class="co">#&gt; CRS:           NA</span></span>
<span><span class="co">#&gt; First 10 features:</span></span>
<span><span class="co">#&gt;                  sample_id                       geometry</span></span>
<span><span class="co">#&gt; AACACACGCTCGCCGC  sample02 POLYGON ((6597.869 7842.575...</span></span>
<span><span class="co">#&gt; AACCGCTAAGGGATGC  sample02 POLYGON ((6724.811 9000, 67...</span></span>
<span><span class="co">#&gt; AACGCTGTTGCTGAAA  sample02 POLYGON ((6457.635 7118.991...</span></span>
<span><span class="co">#&gt; AACGGACGTACGTATA  sample02 POLYGON ((6737.064 8083.571...</span></span>
<span><span class="co">#&gt; AATAGAATCTGTTTCA  sample02 POLYGON ((7570.153 8564.368...</span></span>
<span><span class="co">#&gt; ACAAATCGCACCGAAT  sample02 POLYGON ((8000 7997.001, 79...</span></span>
<span><span class="co">#&gt; ACAATTGTGTCTCTTT  sample02 POLYGON ((6043.169 7843.77,...</span></span>
<span><span class="co">#&gt; ACAGGCTTGCCCGACT  sample02 POLYGON ((7428.88 7358.195,...</span></span>
<span><span class="co">#&gt; ACCAGTGCGGGAGACG  sample02 POLYGON ((6460.753 8566.757...</span></span>
<span><span class="co">#&gt; ACCCTCCCTTGCTATT  sample02 POLYGON ((7847.503 8563.771...</span></span></code></pre></div>
<p>Sample IDs can also be changed, with the <code>changeSampleIDs</code>
function, with a named vector whose names are the old names and values
are the new names.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/changeSampleIDs.html" class="external-link">changeSampleIDs</a></span><span class="op">(</span><span class="va">sfe</span>, replacement <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Vis5A <span class="op">=</span> <span class="st">"foo"</span>, sample02 <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/sampleIDs.html" class="external-link">sampleIDs</a></span><span class="op">(</span><span class="va">sfe_combined</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "foo"</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="object-construction">Object construction<a class="anchor" aria-label="anchor" href="#object-construction"></a>
</h2>
<div class="section level3">
<h3 id="from-scratch">From scratch<a class="anchor" aria-label="anchor" href="#from-scratch"></a>
</h3>
<p>An SFE object can be constructed from scratch with the assay
matrices, column metadata, and spatial coordinates. In this toy example,
<code>dgCMatrix</code> is used, but since SFE inherits from
SingleCellExperiment (SCE), other types of arrays supported by SCE such
as delayed arrays should also work.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visium barcode location from Space Ranger</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"visium_row_col"</span><span class="op">)</span></span>
<span><span class="va">coords1</span> <span class="op">&lt;-</span> <span class="va">visium_row_col</span><span class="op">[</span><span class="va">visium_row_col</span><span class="op">$</span><span class="va">col</span> <span class="op">&lt;</span> <span class="fl">6</span> <span class="op">&amp;</span> <span class="va">visium_row_col</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;</span> <span class="fl">6</span>,<span class="op">]</span></span>
<span><span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">row</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Random toy sparse matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">col_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">13</span>, <span class="fl">13</span><span class="op">)</span></span>
<span><span class="va">row_inds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">13</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="va">row_inds</span>, j <span class="op">=</span> <span class="va">col_inds</span>, x <span class="op">=</span> <span class="va">values</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">coords1</span><span class="op">$</span><span class="va">barcode</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">LETTERS</span>, <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>That should be sufficient to create an SPE object, and an SFE object,
even though no <code>sf</code> data frame was constructed for the
geometries. The constructor behaves similarly to the SPE constructor.
The centroid coordinates of the Visium spots in the toy example can be
converted into spot polygons with the <code>spotDiameter</code>
argument. Spot diameter in pixels in full resolution image can be found
in the <code>scalefactors_json.json</code> file in Space Ranger
output.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colData <span class="op">=</span> <span class="va">coords1</span>,</span>
<span>                                spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>,</span>
<span>                                spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p>More geometries and spatial graphs can be added after calling the
constructor.</p>
<p>Geometries can also be supplied in the constructor.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert regular data frame with coordinates to sf data frame</span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/df2sf.html" class="external-link">df2sf</a></span><span class="op">(</span><span class="va">coords1</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>, <span class="st">"row"</span><span class="op">)</span>, spotDiameter <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cg</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/SpatialFeatureExperiment.html" class="external-link">SpatialFeatureExperiment</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mat</span><span class="op">)</span>, colGeometries <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>foo <span class="op">=</span> <span class="va">cg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="space-ranger-output">Space Ranger output<a class="anchor" aria-label="anchor" href="#space-ranger-output"></a>
</h3>
<p>Space Ranger output can be read in a similar manner as in
<code>SpatialExperiment</code>; the returned SFE object has the
<code>spotPoly</code> column geometry for the spot polygons. If the
filtered matrix is read in, then a column graph called
<code>visium</code> will also be present, for the spatial neighborhood
graph of the Visium spots on tissue. The graph is not computed if all
spots are read in regardless of whether they are on tissue.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Example from SpatialExperiment</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"10xVisium"</span><span class="op">)</span>, </span>
<span>  package <span class="op">=</span> <span class="st">"SpatialExperiment"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"section1"</span>, <span class="st">"section2"</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="va">sample_ids</span>, <span class="st">"outs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The "outs" directory in Space Ranger output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "raw_feature_bc_matrix" "spatial"</span></span>
<span></span>
<span><span class="co"># The contents of the "spatial" directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"spatial"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "scalefactors_json.json"    "tissue_lowres_image.png"  </span></span>
<span><span class="co">#&gt; [3] "tissue_positions_list.csv"</span></span>
<span></span>
<span><span class="co"># Where the raw gene count matrix is located</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">samples</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"raw_feature_bc_matrix"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/usr/local/lib/R/site-library/SpatialExperiment/extdata/10xVisium/section1/outs/raw_feature_bc_matrix"</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/read10xVisiumSFE.html" class="external-link">read10xVisiumSFE</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"raw"</span>,</span>
<span>                         load <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 50 99 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(50): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000005886 ENSMUSG00000101476</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(99): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...</span></span>
<span><span class="co">#&gt;   AAAGTCGACCCTCAGT-1-1 AAAGTGCCATCAATTA-1-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_row_in_fullres pxl_col_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries:  () </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; section1: </span></span>
<span><span class="co">#&gt; section2:</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="coercion-from-spatialexperiment">Coercion from <code>SpatialExperiment</code><a class="anchor" aria-label="anchor" href="#coercion-from-spatialexperiment"></a>
</h3>
<p>SPE objects can be coerced into SFE objects. If column geometries or
spot diameter are not specified, then a column geometry called
“centroids” will be created.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/read10xVisium.html" class="external-link">read10xVisium</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">sample_ids</span>, type <span class="op">=</span> <span class="st">"sparse"</span>, data <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>  images <span class="op">=</span> <span class="st">"lowres"</span>, load <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>For the coercion, column names must not be duplicate.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html" class="external-link">make.unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/SpatialFeatureExperiment-coercion.html" class="external-link">toSpatialFeatureExperiment</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 50 99 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(50): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000005886 ENSMUSG00000101476</span></span>
<span><span class="co">#&gt; rowData names(1): symbol</span></span>
<span><span class="co">#&gt; colnames(99): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...</span></span>
<span><span class="co">#&gt;   AAAGTCGACCCTCAGT-1-1 AAAGTGCCATCAATTA-1-1</span></span>
<span><span class="co">#&gt; colData names(4): in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#&gt; imgData names(4): sample_id image_id data scaleFactor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: centroids (POINT) </span></span>
<span><span class="co">#&gt; annotGeometries:  () </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; section1: </span></span>
<span><span class="co">#&gt; section2:</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="operations">Operations<a class="anchor" aria-label="anchor" href="#operations"></a>
</h2>
<div class="section level3">
<h3 id="non-geometric">Non-geometric<a class="anchor" aria-label="anchor" href="#non-geometric"></a>
</h3>
<p>SFE objects can be concatenated with <code>cbind</code>, as was done
just now to create a toy example with 2 samples.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">sfe1</span>, <span class="va">sfe2</span><span class="op">)</span></span></code></pre></div>
<p>The SFE object can also be subsetted like a matrix, like an SCE
object. More complexity arises when it comes to the spatial graphs. The
<code>drop</code> argument of the SFE method <code>[</code> determines
what to do with the spatial graphs. If <code>drop = TRUE</code>, then
all spatial graphs will be removed, since the graphs with nodes and
edges that have been removed are no longer valid. If
<code>drop = FALSE</code>, which is the default, then the spatial graphs
will be reconstructed with the remaining nodes after subsetting.
Reconstruction would only work when the original graphs were constructed
with <code>findSpatialNeighbors</code> or <code>findVisiumGraph</code>
in this package, which records the method and parameters used to
construct the graphs. If reconstruction fails, then a warning will be
issued and the graphs removed.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe_subset</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; Node indices in the graphs are no longer valid after subsetting. Dropping all row and col graphs.</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 10 10 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(10): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000090031 ENSMUSG00000033740</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(10): AAATTACCTATCGATG AACATATCAACTGGTG ... ACGAGTACGGATGCCC</span></span>
<span><span class="co">#&gt;   ACGCTAGTGATACACT</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON), foobar (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (GEOMETRY), myofiber_simplified (GEOMETRY), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Will give warning because graph reconstruction fails</span></span>
<span><span class="va">sfe_subset</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="geometric">Geometric<a class="anchor" aria-label="anchor" href="#geometric"></a>
</h3>
<p>Just like <code>sf</code> data frames, SFE objects can be subsetted
by a geometry and a predicate relating geometries. For example, if all
Visium spots were read into an SFE object regardless of whether they are
in tissue, and the <code>tissueBoundary</code> annotation geometry is
provided, then the tissue boundary geometry can be used to subset the
SFE object to obtain a new SFE object with only spots on tissue. Loupe
does not give the tissue boundary polygon; such polygon can be obtained
by thresholding the H&amp;E image and converting the mask into polygons
with OpenCV or the <code>terra</code> R package, or by manual annotation
in QuPath or LabKit (the latter needs to be converted into polygon).</p>
<p>Use the <code>crop</code> function to directly get the subsetted SFE
object. Note that in this version of this package, <code>crop</code>
does NOT crop the image.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Before</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"gray"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_SFE_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_in_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">sfe</span>, y <span class="op">=</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p>Note that for large datasets with many geometries, cropping can take
a while to run.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># After</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_geometry.html" class="external-link">st_geometry</a></span><span class="op">(</span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotGeometries.html" class="external-link">tissueBoundary</a></span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe_in_tissue</span><span class="op">)</span>, col <span class="op">=</span> <span class="st">"gray"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig2_SFE_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<p><code>crop</code> can also be used in the conventional sense of
cropping, i.e. specifying a bounding box.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_cropped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span>,</span>
<span>                    xmin <span class="op">=</span> <span class="fl">5500</span>, xmax <span class="op">=</span> <span class="fl">6500</span>, ymin <span class="op">=</span> <span class="fl">13500</span>, ymax <span class="op">=</span> <span class="fl">14500</span><span class="op">)</span></span></code></pre></div>
<p>The <code>colGeometryName</code> is used to determine which columns
in the gene count matrix to keep. All geometries in the SFE object will
be subsetted so only portions intersecting <code>y</code> or the
bounding box are kept. Since the intersection operation can produce a
mixture of geometry types, such as intersection of two polygons
producing polygons, points, and lines, the geometry types of the
<code>sf</code> data frames after subsetting may be different from those
of the originals.</p>
<p>The cropping is done independently for each <code>sample_id</code>,
and only on <code>sample_id</code>s specified. Again,
<code>sample_id</code> is optional when there is only one sample in the
SFE object.</p>
<p>Geometry predicates and operations can also be performed to return
the results without subsetting an SFE object. For example, one may want
a logical vector indicating whether each Visium spot intersects the
tissue, or a numeric vector of how many nuclei there are in each Visium
spot. Or get the intersections between each Visium spot and nuclei.
Again, the geometry predicates and operations are performed
independently for each sample, and the <code>sample_id</code> argument
is optional when there is only one sample.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get logical vector of whether each Visium spot intersects tissue</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotPred.html" class="external-link">annotPred</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                                    annotGeometryName <span class="op">=</span> <span class="st">"tissueBoundary"</span>,</span>
<span>                                    sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span>
<span><span class="co"># Get the number of nuclei per Visium spot</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">n_nuclei</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotPred.html" class="external-link">annotNPred</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="st">"spotPoly"</span>, annotGeometryName <span class="op">=</span> <span class="st">"nuclei"</span><span class="op">)</span></span>
<span><span class="co"># Get geometries of intersections of Visium spots and myofibers</span></span>
<span><span class="va">spot_intersections</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotOp.html" class="external-link">annotOp</a></span><span class="op">(</span><span class="va">sfe</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>, </span>
<span>                              annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<p>Sometimes the spatial coordinates of different samples can take very
different values. The values can be made more comparable by moving all
tissues so the bottom left corner of the bounding box would be at the
origin, which would facilitate plotting and comparison across samples
with <code>geom_sf</code> and <code>facet_*</code>.</p>
<p>To find the bounding box of all geometries in each sample of an SFE
object:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  xmin  ymin  xmax  ymax </span></span>
<span><span class="co">#&gt;  5000 13000  7000 15000</span></span></code></pre></div>
<p>To move the coordinates:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_moved</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/removeEmptySpace.html" class="external-link">removeEmptySpace</a></span><span class="op">(</span><span class="va">sfe</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/bbox-SpatialFeatureExperiment-method.html" class="external-link">bbox</a></span><span class="op">(</span><span class="va">sfe_moved</span>, sample_id <span class="op">=</span> <span class="st">"Vis5A"</span><span class="op">)</span></span>
<span><span class="co">#&gt; xmin ymin xmax ymax </span></span>
<span><span class="co">#&gt;    0    0 2000 2000</span></span></code></pre></div>
<p>The original bounding box before moving is stored within the SFE
object, which can be read by <code>dimGeometry</code> setters so newly
added geometries can have coordinates moved as well; this behavior can
be turned off with the optional argument <code>translate = FALSE</code>
in <code>dimGeometry</code> setters.</p>
</div>
</div>
<div class="section level2">
<h2 id="limitations">Limitations<a class="anchor" aria-label="anchor" href="#limitations"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Only 2D data is supported for now.</li>
<li>Geometric operation do not change the SPE <code>ImgData</code>.</li>
<li>Does not support raster data or <code>spatstat</code> spatial point
process analyses.</li>
</ol>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] Matrix_1.4-1                    sf_1.0-8                       </span></span>
<span><span class="co">#&gt;  [3] SFEData_0.99.0                  SpatialExperiment_1.7.0        </span></span>
<span><span class="co">#&gt;  [5] SingleCellExperiment_1.19.0     SummarizedExperiment_1.27.1    </span></span>
<span><span class="co">#&gt;  [7] Biobase_2.57.1                  GenomicRanges_1.49.0           </span></span>
<span><span class="co">#&gt;  [9] GenomeInfoDb_1.33.3             IRanges_2.31.0                 </span></span>
<span><span class="co">#&gt; [11] S4Vectors_0.35.1                BiocGenerics_0.43.0            </span></span>
<span><span class="co">#&gt; [13] MatrixGenerics_1.9.1            matrixStats_0.62.0             </span></span>
<span><span class="co">#&gt; [15] SpatialFeatureExperiment_0.99.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] rjson_0.2.21                  deldir_1.0-6                 </span></span>
<span><span class="co">#&gt;   [3] ellipsis_0.3.2                class_7.3-20                 </span></span>
<span><span class="co">#&gt;   [5] rprojroot_2.0.3               scuttle_1.7.2                </span></span>
<span><span class="co">#&gt;   [7] XVector_0.37.0                fs_1.5.2                     </span></span>
<span><span class="co">#&gt;   [9] proxy_0.4-27                  bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [11] AnnotationDbi_1.59.1          interactiveDisplayBase_1.35.0</span></span>
<span><span class="co">#&gt;  [13] fansi_1.0.3                   codetools_0.2-18             </span></span>
<span><span class="co">#&gt;  [15] R.methodsS3_1.8.2             sparseMatrixStats_1.9.0      </span></span>
<span><span class="co">#&gt;  [17] cachem_1.0.6                  knitr_1.39                   </span></span>
<span><span class="co">#&gt;  [19] jsonlite_1.8.0                dbplyr_2.2.1                 </span></span>
<span><span class="co">#&gt;  [21] png_0.1-7                     R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;  [23] shiny_1.7.2                   HDF5Array_1.25.1             </span></span>
<span><span class="co">#&gt;  [25] BiocManager_1.30.18           compiler_4.2.0               </span></span>
<span><span class="co">#&gt;  [27] httr_1.4.3                    dqrng_0.3.0                  </span></span>
<span><span class="co">#&gt;  [29] assertthat_0.2.1              fastmap_1.1.0                </span></span>
<span><span class="co">#&gt;  [31] limma_3.53.4                  cli_3.3.0                    </span></span>
<span><span class="co">#&gt;  [33] later_1.3.0                   s2_1.1.0                     </span></span>
<span><span class="co">#&gt;  [35] htmltools_0.5.3               tools_4.2.0                  </span></span>
<span><span class="co">#&gt;  [37] glue_1.6.2                    GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">#&gt;  [39] dplyr_1.0.9                   rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [41] wk_0.6.0                      Rcpp_1.0.9                   </span></span>
<span><span class="co">#&gt;  [43] jquerylib_0.1.4               pkgdown_2.0.6                </span></span>
<span><span class="co">#&gt;  [45] raster_3.5-21                 Biostrings_2.65.1            </span></span>
<span><span class="co">#&gt;  [47] vctrs_0.4.1                   rhdf5filters_1.9.0           </span></span>
<span><span class="co">#&gt;  [49] spdep_1.2-4                   ExperimentHub_2.5.0          </span></span>
<span><span class="co">#&gt;  [51] DelayedMatrixStats_1.19.0     xfun_0.31                    </span></span>
<span><span class="co">#&gt;  [53] stringr_1.4.0                 beachmat_2.13.4              </span></span>
<span><span class="co">#&gt;  [55] mime_0.12                     lifecycle_1.0.1              </span></span>
<span><span class="co">#&gt;  [57] terra_1.5-34                  AnnotationHub_3.5.0          </span></span>
<span><span class="co">#&gt;  [59] edgeR_3.39.3                  zlibbioc_1.43.0              </span></span>
<span><span class="co">#&gt;  [61] promises_1.2.0.1              ragg_1.2.2                   </span></span>
<span><span class="co">#&gt;  [63] parallel_4.2.0                rhdf5_2.41.1                 </span></span>
<span><span class="co">#&gt;  [65] curl_4.3.2                    yaml_2.3.5                   </span></span>
<span><span class="co">#&gt;  [67] memoise_2.0.1                 sass_0.4.2                   </span></span>
<span><span class="co">#&gt;  [69] stringi_1.7.8                 RSQLite_2.2.15               </span></span>
<span><span class="co">#&gt;  [71] BiocVersion_3.16.0            highr_0.9                    </span></span>
<span><span class="co">#&gt;  [73] desc_1.4.1                    e1071_1.7-11                 </span></span>
<span><span class="co">#&gt;  [75] filelock_1.0.2                boot_1.3-28                  </span></span>
<span><span class="co">#&gt;  [77] BiocParallel_1.31.10          spData_2.0.1                 </span></span>
<span><span class="co">#&gt;  [79] rlang_1.0.4                   pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [81] systemfonts_1.0.4             bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;  [83] evaluate_0.15                 lattice_0.20-45              </span></span>
<span><span class="co">#&gt;  [85] purrr_0.3.4                   Rhdf5lib_1.19.2              </span></span>
<span><span class="co">#&gt;  [87] bit_4.0.4                     tidyselect_1.1.2             </span></span>
<span><span class="co">#&gt;  [89] magrittr_2.0.3                R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [91] magick_2.7.3                  generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [93] DelayedArray_0.23.0           DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [95] pillar_1.8.0                  units_0.8-0                  </span></span>
<span><span class="co">#&gt;  [97] KEGGREST_1.37.3               RCurl_1.98-1.7               </span></span>
<span><span class="co">#&gt;  [99] sp_1.5-0                      tibble_3.1.7                 </span></span>
<span><span class="co">#&gt; [101] crayon_1.5.1                  DropletUtils_1.17.0          </span></span>
<span><span class="co">#&gt; [103] KernSmooth_2.23-20            utf8_1.2.2                   </span></span>
<span><span class="co">#&gt; [105] BiocFileCache_2.5.0           rmarkdown_2.14               </span></span>
<span><span class="co">#&gt; [107] locfit_1.5-9.6                grid_4.2.0                   </span></span>
<span><span class="co">#&gt; [109] blob_1.2.3                    digest_0.6.29                </span></span>
<span><span class="co">#&gt; [111] classInt_0.4-7                xtable_1.8-4                 </span></span>
<span><span class="co">#&gt; [113] httpuv_1.6.5                  R.utils_2.12.0               </span></span>
<span><span class="co">#&gt; [115] textshaping_0.3.6             bslib_0.4.0</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Lambda Moses, Lior Pachter.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
