<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>3. Voyager • SFEWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="3. Voyager">
<meta property="og:description" content="SFEWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SFEWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vig1_intro.html">1. From geospatial to spatial transcriptomics</a>
    </li>
    <li>
      <a href="../articles/vig2_SFE.html">2. SpatialFeatureExperiment</a>
    </li>
    <li>
      <a href="../articles/vig3_Voyager.html">3. Voyager</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lambdamoses/SFEWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>3. Voyager</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/lambdamoses/SFEWorkshop/blob/HEAD/vignettes/vig3_Voyager.Rmd"><code>vignettes/vig3_Voyager.Rmd</code></a></small>
      <div class="hidden name"><code>vig3_Voyager.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Voyager</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/dev/" class="external-link">SpatialFeatureExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SingleCellExperiment</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span></span>
<span><span class="co">#&gt;     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span></span>
<span><span class="co">#&gt;     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span><span class="co">#&gt;     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span></span>
<span><span class="co">#&gt;     union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span><span class="co">#&gt; Loading required package: scuttle</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SFEData</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1; sf_use_s2() is TRUE</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scales.r-lib.org" class="external-link">scales</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/const-ae/sparseMatrixStats" class="external-link">sparseMatrixStats</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In single cell RNA-seq (scRNA-seq), data and metadata can be
represented with <code>SingleCellExperiment</code> or
<code>Seurat</code> objects, and basic exploratory data analyses and
visualization performed with <code>scater</code>, <code>scran</code>,
and <code>scuttle</code>, or <code>Seurat</code>. The
<code>SpatialFeatureExperiment</code> package and S4 class extending
<code>SpatialExperiment</code> and <code>SingleCellExperiment</code>
brings EDA methods for vector spatial data to spatial transcriptomics.
<code>Voyager</code> to <code>SpatialFeatureExperiment</code> is just
like <code>scater</code>, <code>scran</code>, and <code>scuttle</code>
to <code>SingleCellExperiment</code>, implementing basic exploratory
spatial data analysis (ESDA) and visualization methods. Non-spatial
statistical methods often assume that the samples (cells, spots) are
independent, which is not the case in spatial data, where nearby samples
tend to be more similar (i.e. spatial autocorrelation). Much of ESDA is
dedicated to spatial autocorrelation, such as finding whether it is
present, and if so what’s its length scale.</p>
<p>This vignette gives an overview of these spatial EDA methods,
functionalities of the <code>Voyager</code> package, and applications of
the <code>SpatialFeatureExperiment</code> class with a published Visium
dataset.</p>
</div>
<div class="section level2">
<h2 id="dataset">Dataset<a class="anchor" aria-label="anchor" href="#dataset"></a>
</h2>
<p>The dataset used in this vignette comes from <a href="https://doi.org/10.1038/s42003-021-02810-x" class="external-link">Large-scale
integration of single-cell transcriptomic data captures transitional
progenitor states in mouse skeletal muscle regeneration</a>. Notexin was
injected into the tibialis anterior muscle to induce injury, and the
healing muscle was collected 2, 5, and 7 days post injury for Visium.
The dataset here is from the 2 day timepoint. The dataset is in a
<code>SpatialFeatureExperiment</code> (SFE) object.</p>
<p>The gene count matrix was directly downloaded from GEO. All 4992
spots, whether in tissue or not, are included. The H&amp;E image was
used for nuclei and myofiber segmentation. A subset of nuclei from
randomly selected regions from all 3 timepoints were manually annotated
to train a StarDist model to segment the rest of the nuclei, and the
myofibers were all manually segmented.</p>
<p>Tissue boundary, nuclei, myofiber, and Visium spot polygons are
stored as <code>sf</code> data frames in the SFE object. See the
vignette of the <code>SpatialFeatureExperiment</code> for more details
on the structure of the SFE object. The SFE object of this dataset is
provided in the <code>SFEData</code> package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">sfe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SFEData/man/McKellarMuscleData.html" class="external-link">McKellarMuscleData</a></span><span class="op">(</span><span class="st">"full"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2022-07-15</span></span>
<span><span class="co">#&gt; see ?SFEData and browseVignettes('SFEData') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; class: SpatialFeatureExperiment </span></span>
<span><span class="co">#&gt; dim: 15123 4992 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(15123): ENSMUSG00000025902 ENSMUSG00000096126 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064368 ENSMUSG00000064370</span></span>
<span><span class="co">#&gt; rowData names(6): Ensembl symbol ... vars cv2</span></span>
<span><span class="co">#&gt; colnames(4992): AAACAACGAATAGTTC AAACAAGTATCTCCCA ... TTGTTTGTATTACACG</span></span>
<span><span class="co">#&gt;   TTGTTTGTGTAAATTC</span></span>
<span><span class="co">#&gt; colData names(12): barcode col ... prop_mito in_tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span><span class="co">#&gt; spatialCoords names(2) : imageX imageY</span></span>
<span><span class="co">#&gt; imgData names(1): sample_id</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Geometries:</span></span>
<span><span class="co">#&gt; colGeometries: spotPoly (POLYGON) </span></span>
<span><span class="co">#&gt; annotGeometries: tissueBoundary (POLYGON), myofiber_full (POLYGON), myofiber_simplified (POLYGON), nuclei (POLYGON), nuclei_centroid (POINT) </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Graphs:</span></span>
<span><span class="co">#&gt; Vis5A:</span></span></code></pre></div>
The H&amp;E image of this section:
<div class="figure" style="text-align: center">
<img src="tissue_lowres_5a.jpeg" alt="A cross section of mouse muscle is slightly off center to the lower left. In the middle of the tissue is the notexin injury site with leukocyte infiltration and fewer myofibers. The rest of the tissue section is tightly packed with myofibers." width="100%"><p class="caption">
Low resolution H&amp;E image of the tissue section
</p>
</div>
</div>
<div class="section level2">
<h2 id="exploratory-data-analysis">Exploratory data analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h2>
<div class="section level3">
<h3 id="spots-in-tissue">Spots in tissue<a class="anchor" aria-label="anchor" href="#spots-in-tissue"></a>
</h3>
<p>While the example dataset has all Visium spots whether on tissue or
not, only spots that intersect tissue will be used for further
analyses.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "barcode"   "col"       "row"       "x"         "y"         "dia"      </span></span>
<span><span class="co">#&gt;  [7] "tissue"    "sample_id" "nCounts"   "nGenes"    "prop_mito" "in_tissue"</span></span></code></pre></div>
<p>Total UMI counts (<code>nCounts</code>), number of genes detected per
spot (<code>nGenes</code>), and proportion of mitochondrially encoded
counts (<code>prop_mito</code>) have been precomputed and are in
<code>colData(sfe)</code>. The <code>plotSpatialFeature</code> function
plots any gene, <code>colData</code> values, and geometry attributes in
<code>colGeometry</code> and <code>annotGeometry</code> in space. The
Visium spots are plotted as polygons reflecting their actual size
relative to the tissue, rather than points as in other packages that
plot Visium data. Behind the scene, <code>geom_sf</code> is used to plot
the geometries.</p>
<p>The tissue boundary was found by thresholding the H&amp;E image and
removing small polygons that are most likely debris. The
<code>in_tissue</code> column of <code>colData(sfe)</code> indicates
which Visium spot polygon intersects the tissue polygon; this can be
found with <code><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotPred.html" class="external-link">SpatialFeatureExperiment::annotPred()</a></code> but is
beyond the scope of this vignette.</p>
<p>While <code>scran</code> is used for data normalization here for
demonstration purposes and to make the data more normally distributed,
we do not mean that it is the best practice in normalizing spatial
transcriptomics data, as we don’t know what the best practice really
should be. As seen in the <code>nCounts</code> plot in space above,
spatial autocorrelation is evident. In Visium, reverse transcription
occurs in situ on the spots, but PCR amplification occurs after the cDNA
is dissociated from the spots. Then artifacts introduced from the
amplification step would not be spatial. Spatial artifacts may arise
from diffusion of transcripts to adjacent spots and tissue
permeablization. However, given how the total counts seem to correspond
to histological regions, the total counts may have a biological
component and hence should not be treated as a technical artifact to be
normalized away as in scRNA-seq data normalization methods.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe</span><span class="op">[</span>,<span class="fu">colData</span><span class="op">(</span><span class="va">sfe</span><span class="op">)</span><span class="op">$</span><span class="va">in_tissue</span><span class="op">]</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="va">sfe_tissue</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>,<span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, clusters<span class="op">=</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in (function (x, sizes, min.mean = NULL, positive = FALSE, scaling =</span></span>
<span><span class="co">#&gt; NULL) : encountered non-positive size factor estimates</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span></code></pre></div>
<p>Myofiber and nuclei segmentation polygons are available in this
dataset, in the field <code>annotGeometries</code>. Myofibers were
manually segmented, and nuclei were segmented with <a href="https://github.com/stardist/stardist" class="external-link"><code>StarDist</code></a>,
trained with a manually segmented subset.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotGeometries.html" class="external-link">annotGeometryNames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "tissueBoundary"      "myofiber_full"       "myofiber_simplified"</span></span>
<span><span class="co">#&gt; [4] "nuclei"              "nuclei_centroid"</span></span></code></pre></div>
<div class="section level4">
<h4 id="from-myofibers-and-nuclei-to-visium-spots">From myofibers and nuclei to Visium spots<a class="anchor" aria-label="anchor" href="#from-myofibers-and-nuclei-to-visium-spots"></a>
</h4>
<p>The <code>plotSpatialFeature</code> function can also be used to plot
attributes of geometries, i.e. the non-geometry columns in the
<code>sf</code> data frames in the <code>rowGeometries</code>,
<code>colGeometries</code>, or <code>annotGeometries</code> fields in
the SFE object. For <code>rowGeometries</code> and
<code>colGeometries</code>, such columns associated with the
<code>sf</code> data frames rather than <code>rowData</code> or
<code>colData</code> are allowed because one can specify how these
columns associate with the geometries (see <a href="https://r-spatial.github.io/sf/reference/st_agr.html" class="external-link"><code>st_agr</code></a>
and <a href="https://r-spatial.github.io/sf/reference/sf.html#details-1" class="external-link">documentation
of <code>st_sf</code></a>). When an attribute of an
<code>annotGeometry</code> is plotted along side gene expression or
<code>colData</code> or <code>colGeometry</code> attribute, the
<code>annotGeometry</code> attribute is plotted with a different color
palette to distinguish from the column associated values.</p>
<p>Here, from the <code>annotGeometries</code>, the myofiber polygons
are plotted, colored by cross section area as observed in this tissue
section. The <code>aes_use</code> argument is set to <code>color</code>
rather than <code>fill</code> (default for polygons) to only plot the
Visium spot outlines to make the myofiber polygons more visible. The
<code>fill</code> argument is set to <code>NA</code> to make the Visium
spots look hollow, and the <code>size</code> argument controls the
thickness of the outlines. The <code>annot_aes</code> argument specifies
which column in the <code>annotGeometry</code> to use to specify the
values of an aesthstic, just like <code>aes</code> in
<code>ggplot2</code> (<code>aes_string</code> to be precise, since
<code>tidyeval</code> is not used here). The <code>annot_fixed</code>
argument (not used here) can set the fixed size, alpha, color, and etc.
for the <code>annotGeometry</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="st">"nCounts"</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, </span>
<span>                   aes_use <span class="op">=</span> <span class="st">"color"</span>, size <span class="op">=</span> <span class="fl">0.5</span>, fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-8-1.png" title="Plot of Visium spots in tissue and myofiber polygons in physical space. Visium spots are colored by nCounts, and myofibers are colored by area." alt="Plot of Visium spots in tissue and myofiber polygons in physical space. Visium spots are colored by nCounts, and myofibers are colored by area." width="700" style="display: block; margin: auto;"></p>
<p>The larger myofibers seem to have fewer total counts, possibly
because the larger size of these myofibers dilute the transcripts. If
this is the case, then data normalization would be relevant to correct
for this.</p>
<p>With <code>SpatialFeatureExperiment</code>, we can find the number of
myofibers and nuclei that intersect each Visium spot. The predicate can
be <a href="https://r-spatial.github.io/sf/reference/geos_binary_pred.html" class="external-link">anything
implemented in <code>sf</code></a>, so for example, the number of nuclei
fully covered by each Visium spot can also be found. The default
predicate is <code>st_intersects</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">n_myofibers</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotPred.html" class="external-link">annotNPred</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>             annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<p>We can also find the intersection geometries between each myofiber
and Visium spot, and compute their areas.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myofiber_in_spot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotOp.html" class="external-link">annotOp</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span>,</span>
<span>                            annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">prop_myofiber</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_measures.html" class="external-link">st_area</a></span><span class="op">(</span><span class="va">myofiber_in_spot</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://r-spatial.github.io/sf/reference/geos_measures.html" class="external-link">st_area</a></span><span class="op">(</span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/dimGeometries.html" class="external-link">spotPoly</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n_myofibers"</span>, <span class="st">"prop_myofibers"</span><span class="op">)</span>, </span>
<span>                   colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-11-1.png" title="Plot of Visium spots in tissue in physical space, colored by number of myofibers intersecting each spot and proportion of area of each spot occupied by myofibers." alt="Plot of Visium spots in tissue in physical space, colored by number of myofibers intersecting each spot and proportion of area of each spot occupied by myofibers." width="768" style="display: block; margin: auto;"></p>
<p>The injury site infiltrated with immune cells have more nuclei; this
is apparent from the H&amp;E image, but is quantified here.</p>
<p>There is no one to one mapping between Visium spots and myofibers.
However, we may want to relate attributes of myofibers to gene
expression detected at the Visium spots. One way to do so is to
summarize the attributes of all myofibers that intersect (or choose
another better predicate implemented in <code>sf</code>) each spot, such
as to calculate the mean, median, or sum. This can be done with the
<code>annotSummary</code> function in
<code>SpatialFeatureExperiment</code>. The default predicate is
<code>st_intersects</code>, and the default summary function is
<code>mean</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">mean_myofiber_area</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotSummary.html" class="external-link">annotSummary</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"spotPoly"</span>, <span class="st">"myofiber_simplified"</span>, </span>
<span>               annotColNames <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="co"># it always returns a data frame</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The gray spots don't intersect any myofiber</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"mean_myofiber_area"</span>, <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-13-1.png" title="Plot of Visium spots in tissue in physical space, colored by the average area of myofibers that intersect each spot. The average area is higher near the mid-top right part of the tissue." alt="Plot of Visium spots in tissue in physical space, colored by the average area of myofibers that intersect each spot. The average area is higher near the mid-top right part of the tissue." width="700" style="display: block; margin: auto;"></p>
<p>Now we can see how the mean area of myofibers intersecting each
Visium spot relates to other aspects of the spots such as total counts
and gene expression.</p>
<p>The NAs are for spots not intersecting any myofibers, e.g. those in
the inflammatory region. There seem to be two clusters: spots with
smaller mean myofiber area and larger total counts, and spots with
larger mean myofiber area and smaller total counts. Indeed, spots with
smaller mean myofiber area tend to have a larger number of myofibers;
that the myofibers are smaller is the reason why there are more of them
intersecting each spot.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, colour_by <span class="op">=</span> <span class="st">"mean_myofiber_area"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-14-1.png" title="Scatter plot with nGenes in the y axis and nCounts in the x axis. Each point represents a Visium spot in tissue, colored by mean myofiber area." alt="Scatter plot with nGenes in the y axis and nCounts in the x axis. Each point represents a Visium spot in tissue, colored by mean myofiber area." width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"nGenes"</span>, colour_by <span class="op">=</span> <span class="st">"n_myofibers"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-15-1.png" title="Scatter plot with nGenes in the y axis and nCounts in the x axis. Each point represents a Visium spot in tissue, colored by number of myofibers." alt="Scatter plot with nGenes in the y axis and nCounts in the x axis. Each point represents a Visium spot in tissue, colored by number of myofibers." width="700" style="display: block; margin: auto;"></p>
<p>Mean myofiber area and number of myofibers per spot seem to be
related to the two branches in the nGenes vs. nCounts plot. The upper
branch includes the inflamed injury site with lots of immune cells and
hence nuclei and devoid of myofibers and the adjacent region with larger
myofibers.</p>
</div>
<div class="section level4">
<h4 id="myofiber-types">Myofiber types<a class="anchor" aria-label="anchor" href="#myofiber-types"></a>
</h4>
<p>Marker genes: Myh7 (Type I, slow twitch, aerobic), Myh2 (Type IIa,
fast twitch, somewhat aerobic), Myh4 (Type IIb, fast twitch, anareobic),
Myh1 (Type IIx, fast twitch, anaerobic), from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5526362/" class="external-link">this
protocol</a></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>I <span class="op">=</span> <span class="st">"Myh7"</span>, IIa <span class="op">=</span> <span class="st">"Myh2"</span>, IIb <span class="op">=</span> <span class="st">"Myh4"</span>, IIx <span class="op">=</span> <span class="st">"Myh1"</span><span class="op">)</span></span></code></pre></div>
<p>First look at Type I myofibers. Again, this is a fast twitch muscle,
so we don’t expect many slow twitch Type I myofibers. Row names in
<code>sfe_tissue</code> are Ensembl IDs, to avoid ambiguity as sometimes
multiple Ensembl IDs have the same gene symbol and some genes have
aliases in symbol. However, gene symbols are shorter and more human
readable than Ensembl IDs, and are nice to show on plots. In the
<code>plotSpatialFeature</code> function and other functions in
<code>Voyager</code>, even when the actual row names are Ensembl IDs,
the <code>features</code> argument can take gene symbols if there is a
column called “symbols” in <code>rowData(sfe)</code>, where the function
converts the gene symbols to Ensembl IDs. By default, gene symbols are
shown on the plot, but the <code>show_symbol</code> argument can be set
to <code>FALSE</code> to show Ensembl IDs instead. If one gene symbol
matches multiple Ensembl IDs in the dataset, then a warning will be
given.</p>
<p>The <code>exprs_values</code> argument specifies the assay to use,
which is by default “logcounts”, i.e. the log normalized data. Again,
this default may or may not be best practice given that total UMI counts
may have biological relevance in spatial data. Therefore we are plotting
both the raw counts and the log normalized counts here.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function specific for this vignette, with some hard coded values</span></span>
<span><span class="va">plot_counts_logcounts</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.5</span>, show_symbol <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                   exprs_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Raw counts"</span><span class="op">)</span></span>
<span>  <span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe</span>, <span class="va">feature</span>, <span class="st">"spotPoly"</span>,</span>
<span>                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, </span>
<span>                   annot_aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"area"</span><span class="op">)</span>, aes_use <span class="op">=</span> <span class="st">"color"</span>,</span>
<span>                   fill <span class="op">=</span> <span class="cn">NA</span>, size <span class="op">=</span> <span class="fl">0.5</span>, show_symbol <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                   exprs_values <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Log normalized counts"</span><span class="op">)</span></span>
<span>  <span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="va">feature</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_counts_logcounts</span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">markers</span><span class="op">[</span><span class="st">"I"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Type IIa myofibers</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_counts_logcounts</span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">markers</span><span class="op">[</span><span class="st">"IIa"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-19-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Type IIa myofibers also tend to be clustered together on left side of
the tissue.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span> <span class="op">==</span> <span class="va">markers</span><span class="op">[</span><span class="st">"IIa"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, </span>
<span>            colour_by <span class="op">=</span> <span class="va">gene_id</span>, by_exprs_values <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-20-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Spots with Type IIa myofibers tend to have higher proportion of
mitochondrial counts, mostly in the upper cluster in this plot. However,
some of these spots are in the lower cluster. As Visium spots can
intersect with multiple myofibers, these spots might have both Type IIa
and some more anaerobic myofibers.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"mean_myofiber_area"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, </span>
<span>            colour_by <span class="op">=</span> <span class="va">gene_id</span>, by_exprs_values <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 36 rows containing missing values (geom_point).</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-21-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Plotting proportion of mitochondrial counts vs. mean myofiber area,
we also see two clusters, one with higher proportion of mitochondrial
counts and smaller area, and another with lower proportion of
mitochondrial counts and on average slightly larger area. Type IIa
myofibers tend to have small area and larger proportion of mitochondrial
counts.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plot_counts_logcounts</span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">markers</span><span class="op">[</span><span class="st">"IIb"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-22-1.png" width="700" style="display: block; margin: auto;"></p>
<p>While the raw counts suggest that Type IIb myofibers tend to
aggregate at the top region of this section and away from Type IIa
myofibers, data normalization made expression of Myh4 seem ubiquitous
even in regions without myofibers.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">symbol</span> <span class="op">==</span> <span class="va">markers</span><span class="op">[</span><span class="st">"IIb"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"nCounts"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, </span>
<span>            colour_by <span class="op">=</span> <span class="va">gene_id</span>, by_exprs_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-23-1.png" width="700" style="display: block; margin: auto;"></p>
<p>In raw counts, spots with high Myh4 expression, indicating Type IIb,
also tend to have somewhat high mitochondrial count proportion (still in
the upper cluster), but not as high as Type IIa.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, x <span class="op">=</span> <span class="st">"mean_myofiber_area"</span>, y <span class="op">=</span> <span class="st">"prop_mito"</span>, </span>
<span>            colour_by <span class="op">=</span> <span class="va">gene_id</span>, by_exprs_values <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 36 rows containing missing values (geom_point).</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-24-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Type IIb myofibers tend to be larger than type IIa myofibers.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="spatial-neighborhood-graphs">Spatial neighborhood graphs<a class="anchor" aria-label="anchor" href="#spatial-neighborhood-graphs"></a>
</h2>
<p>A spatial neighborhood graph is required to compute spatial
dependency metrics such as Moran’s I and Geary’s C. The
<code>SpatialFeatureExperiment</code> package wraps methods in
<code>spdep</code> to find spatial neighborhood graphs, which are stored
within the SFE object (see <code>spdep</code> documentation for <a href="https://r-spatial.github.io/spdep/reference/graphneigh.html" class="external-link"><code>gabrielneigh</code></a>,
<a href="https://r-spatial.github.io/spdep/reference/knearneigh.html" class="external-link"><code>knearneigh</code></a>,
<a href="https://r-spatial.github.io/spdep/reference/poly2nb.html" class="external-link"><code>poly2nb</code></a>,
and <a href="https://r-spatial.github.io/spdep/reference/tri2nb.html" class="external-link"><code>tri2nb</code></a>).
The <code>Voyager</code> package then uses these graphs for spatial
dependency analyses, again based on <code>spdep</code> in this first
version, but methods from other geospatial packages, some of which also
use the spatial neighborhood graphs, may be added later as needed.</p>
<p>For Visium, where the spots are in a hexagonal grid, the spatial
neighborhood graph is straightforward. However, for spatial technologies
with single cell resolution (e.g. MERFISH) and in this dataset, the
myofibers and nuclei, many different methods can be used to find the
spatial neighborhood graph. Here for myofibers, the method “poly2nb”
identifies myofiber polygons that are physically touch each other.
<code>zero.policy = TRUE</code> will allow singletons or nodes without
neighbors in the graph; in the inflamed region, there are more
singletons. We have not yet benchmarked which spatial neighborhood
method is the “best” in which situation; the particular method used here
is for demonstration purpose and may or may not be best practice.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/findVisiumGraph.html" class="external-link">findVisiumGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">annotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_poly2nb"</span><span class="op">)</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/findSpatialNeighbors-SpatialFeatureExperiment-method.html" class="external-link">findSpatialNeighbors</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, type <span class="op">=</span> <span class="st">"myofiber_simplified"</span>, MARGIN <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       method <span class="op">=</span> <span class="st">"poly2nb"</span>, zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The <code>plotColGraph</code> function plots the graph in space
associated with a <code>colGeometry</code>, along with the geometry of
interest.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotColGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-26-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Similarly, the <code>plotAnnotGraph</code> function plots the graph
associated with an <code>annotGeometry</code>, along with the geometry
of interest.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotColGraph.html" class="external-link">plotAnnotGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>               annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-27-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There is no <code>plotRowGraph</code> yet since we haven’t worked
with a dataset where spatial graphs related to genes are relevant,
although the SFE object supports row graphs.</p>
</div>
<div class="section level2">
<h2 id="exploratory-spatial-data-analysis">Exploratory <em>spatial</em> data analysis<a class="anchor" aria-label="anchor" href="#exploratory-spatial-data-analysis"></a>
</h2>
<p>Tobler’s first law of geography:</p>
<blockquote>
<p>Everything is related to everything else. But near things are more
related than distant things.</p>
</blockquote>
<p>All spatial autocorrelation metrics in this package can be computed
directly on a vector or a matrix rather than an SFE object. The user
interface emulates those of dimension reductions in the
<code>scater</code> package (e.g. <code>calculateUMAP</code> that takes
in a matrix or SCE object and returns a matrix, and <code>runUMAP</code>
that takes in an SCE object and adds the results to the
<code>reducedDims</code> field of the SCE object). So
<code>calculate*</code> functions take in a matrix or an SFE object and
directly return the results (format of the results depends on the
structure of the results), while <code>run*</code> functions take in an
SFE object and add the results to the object.</p>
<div class="section level3">
<h3 id="univariate">Univariate<a class="anchor" aria-label="anchor" href="#univariate"></a>
</h3>
<p>In this first version, <code>Voyager</code> only supports univariate
global spatial autocorrelation implemented in <code>spdep</code> for
ESDA: Moran’s I and Geary’s C, permutation testing for Moran’s I and
Geary’s C, Moran plot, and correlograms. In addition, beyond
<code>spdep</code>, <code>Voyager</code> can cluster Moran plots and
correlograms. Plotting functions taking in SFE objects are implemented
to plot the results with <code>ggplot2</code> and with more
customization options than <code>spdep</code> plotting functions.</p>
<p>To demonstrate spatial autocorrelation in gene expression, top highly
variable genes (HVGs) are used. The HVGs are found with the
<code>scran</code> method.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span></span>
<span><span class="va">hvgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="morans-i">Moran’s I<a class="anchor" aria-label="anchor" href="#morans-i"></a>
</h4>
<p>There are several ways to quantify spatial autocorrelation, the most
common of which is Moran’s I:</p>
<p><span class="math display">\[
I = \frac{n}{\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij} (x_i - \bar{x})(x_j - \bar{x})}{\sum_{i=1}^n (x_i -
\bar{x})^2}
\]</span></p>
<p>Where <span class="math inline">\(n\)</span> is the number of spots
or locations, <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are different locations, or spots in
the Visium context, <span class="math inline">\(x\)</span> is a variable
with values at each location, and <span class="math inline">\(w_{ij}\)</span> is a spatial weight, which can be
inversely proportional to distance between spots or an indicator of
whether two spots are neighbors, subject to various definitions of
neighborhood and whether to normalize the number of neighbors. The
<code>spdep</code> package uses the neighborhood.</p>
<p>Moran’s I takes values between -1 and 1. For positive spatial
autocorrelation, i.e. nearby spots tend to be more similar, Moran’s I
will be positive. For negative spatial autocorrelation, i.e. nearby
spots tend to be more dissimilar, Moran’s I will be negative. When the
variable is distributed in space randomly like salt and pepper, then
Moran’s I will be around 0. Positive Moran’s I indicates global
structure, while negative Moran’s I indicates local structure.</p>
<p>Upon visual inspection, total UMI counts per spot seem to have
spatial autocorrelation. A spatial neighborhood graph is required to
compute Moran’s I, and is specified with the <code>listw</code>
argument. Also, if the metrics are computed for a large number of
features, parallel computing is supported, with
<code>BiocParallel</code>, with the <code>BPPARAM</code> argument.</p>
<p>For matrices, the rows are the features, as in the gene count
matrix.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directly use vector or matrix, and multiple features can be specified at once</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoransI.html" class="external-link">calculateMoransI</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>, </span>
<span>                 listw <span class="op">=</span> <span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/spatialGraphs.html" class="external-link">colGraph</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"visium"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 2 columns</span></span>
<span><span class="co">#&gt;                 I         K</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; nCounts  0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes   0.384028   3.88036</span></span></code></pre></div>
<p>I is Moran’s I, and K is sample kurtosis.</p>
<p>To add the results to the SFE object, specifically for colData:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoransI.html" class="external-link">colDataMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,</span>
<span>                             colGraphName <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 2 columns</span></span>
<span><span class="co">#&gt;           MoransI_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; barcode              NA        NA</span></span>
<span><span class="co">#&gt; col                  NA        NA</span></span>
<span><span class="co">#&gt; row                  NA        NA</span></span>
<span><span class="co">#&gt; x                    NA        NA</span></span>
<span><span class="co">#&gt; y                    NA        NA</span></span>
<span><span class="co">#&gt; dia                  NA        NA</span></span>
<span><span class="co">#&gt; tissue               NA        NA</span></span>
<span><span class="co">#&gt; sample_id            NA        NA</span></span>
<span><span class="co">#&gt; nCounts        0.528705   3.00082</span></span>
<span><span class="co">#&gt; nGenes         0.384028   3.88036</span></span></code></pre></div>
<p>For <code>colData</code>, the results are added to
<code>colFeatureData(sfe)</code>, and features for which Moran’s I is
not calculated have NA. The column names of <code>featureData</code>
distinguishes between different samples (there’s only one sample in this
dataset), and are parsed by plotting functions.</p>
<p>To add the results to the SFE object, specifically for geometries:
Here “area” is the area of the cross section of each myofiber as seen in
this tissue section and “eccentricity” is the eccentricity of the
ellipse fitted to each myofiber.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remember zero.policy = TRUE since there're singletons</span></span>
<span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoransI.html" class="external-link">annotGeometryMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, </span>
<span>                                   features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="st">"eccentricity"</span><span class="op">)</span>, </span>
<span>                                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                                   annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>                                   zero.policy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="fu"><a href="https://pachterlab.github.io/SpatialFeatureExperiment/reference/annotGeometries.html" class="external-link">annotGeometry</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"myofiber_simplified"</span><span class="op">)</span>, <span class="st">"featureData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">#&gt;              MoransI_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; lyr.1                   NA        NA</span></span>
<span><span class="co">#&gt; area              0.327888   4.95675</span></span>
<span><span class="co">#&gt; perimeter               NA        NA</span></span>
<span><span class="co">#&gt; eccentricity      0.110938   3.26913</span></span>
<span><span class="co">#&gt; theta                   NA        NA</span></span>
<span><span class="co">#&gt; sine_theta              NA        NA</span></span></code></pre></div>
<p>For a non-geometry column in a <code>colGeometry</code>,
<code>colGeometryMoransI</code> is like
<code>annotGeometryMoransI</code> here, but none of the
<code>colGeometries</code> in this dataset has extra columns.</p>
<p>For gene expression, the <code>logcounts</code> assay is used by
default (use the <code>exprs_values</code> argument to change the
assay), though this may or may not be best practice.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoransI.html" class="external-link">runMoransI</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, </span>
<span>                         BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hvgs</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 8 columns</span></span>
<span><span class="co">#&gt;                               Ensembl      symbol            type     means</span></span>
<span><span class="co">#&gt;                           &lt;character&gt; &lt;character&gt;     &lt;character&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893 ENSMUSG00000018893          Mb Gene Expression   2.11118</span></span>
<span><span class="co">#&gt; ENSMUSG00000027559 ENSMUSG00000027559        Car3 Gene Expression   2.32632</span></span>
<span><span class="co">#&gt; ENSMUSG00000056328 ENSMUSG00000056328        Myh1 Gene Expression   4.82572</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304 ENSMUSG00000029304        Spp1 Gene Expression   1.63722</span></span>
<span><span class="co">#&gt; ENSMUSG00000033196 ENSMUSG00000033196        Myh2 Gene Expression   0.97476</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335 ENSMUSG00000050335      Lgals3 Gene Expression   1.43189</span></span>
<span><span class="co">#&gt;                         vars       cv2 MoransI_Vis5A   K_Vis5A</span></span>
<span><span class="co">#&gt;                    &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000018893   74.1782   16.6428      0.761522   1.80210</span></span>
<span><span class="co">#&gt; ENSMUSG00000027559   74.3233   13.7336      0.706457   1.76082</span></span>
<span><span class="co">#&gt; ENSMUSG00000056328  302.2385   12.9785      0.727282   2.14464</span></span>
<span><span class="co">#&gt; ENSMUSG00000029304   60.1583   22.4430      0.656386   1.72563</span></span>
<span><span class="co">#&gt; ENSMUSG00000033196   24.0374   25.2984      0.706747   2.45049</span></span>
<span><span class="co">#&gt; ENSMUSG00000050335   48.0739   23.4471      0.652904   1.93371</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gearys-c">Geary’s C<a class="anchor" aria-label="anchor" href="#gearys-c"></a>
</h4>
<p>Another spatial autocorrelation metric is Geary’s C, defined as:</p>
<p><span class="math display">\[
C = \frac{(n-1)}{2\sum_{i=1}^n \sum_{j=1}^n w_{ij}} \frac{\sum_{i=1}^n
\sum_{j=1}^n w_{ij}(x_i - x_j)^2}{{\sum_{i=1}^n (x_i - \bar{x})^2}}
\]</span></p>
<p>Geary’s C well below 1 indicates positive spatial autocorrelation,
and above 1 indicates negative spatial autocorrelation.</p>
<p>Simply substitute “MoransI” in the names of the functions in the
previous section with “GearysC” to compute Geary’s C for features of
interest and add the results to the SFE object. For example, for
<code>colData</code></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoransI.html" class="external-link">colDataGearysC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>,</span>
<span>                             colGraphName <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 3 columns</span></span>
<span><span class="co">#&gt;           MoransI_Vis5A   K_Vis5A GearysC_Vis5A</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; barcode              NA        NA            NA</span></span>
<span><span class="co">#&gt; col                  NA        NA            NA</span></span>
<span><span class="co">#&gt; row                  NA        NA            NA</span></span>
<span><span class="co">#&gt; x                    NA        NA            NA</span></span>
<span><span class="co">#&gt; y                    NA        NA            NA</span></span>
<span><span class="co">#&gt; dia                  NA        NA            NA</span></span>
<span><span class="co">#&gt; tissue               NA        NA            NA</span></span>
<span><span class="co">#&gt; sample_id            NA        NA            NA</span></span>
<span><span class="co">#&gt; nCounts        0.528705   3.00082      0.474892</span></span>
<span><span class="co">#&gt; nGenes         0.384028   3.88036      0.605797</span></span></code></pre></div>
<p>There’s only one column for K since it’s the same for Moran’s I and
Geary’s C. Here both Moran’s I and Geary’s C suggest positive spatial
autocorrelation for <code>nCounts</code> and <code>nGenes</code>.</p>
</div>
<div class="section level4">
<h4 id="permutation-testing">Permutation testing<a class="anchor" aria-label="anchor" href="#permutation-testing"></a>
</h4>
<p>Is the spatial autocorrelation statistically significant? The
<code>moran.test</code> function in <code>spdep</code> can give an
analytic p-value but the p-value would not be accurate if the data is
not normally distributed. As gene expression data is generally not
normally distributed and data normalization doesn’t necessarily make the
data that close to a normal distribution, permutation testing is used in
this package to test the significance of Moran’s I and Geary’s C,
wrapping <a href="https://r-spatial.github.io/spdep/reference/moran.mc.html" class="external-link"><code>moran.mc</code>
in <code>spdep</code></a>. Just like Moran’s I, there’s
<code>calculateMoranMC</code> and <code>calculateGearyMC</code>
functions to directly return the results, and
<code>colDataMoranMC</code>, <code>colGeometryMoranMC</code>,
<code>annotGeometryMoranMC</code>, and <code>runMoranMC</code> to add
the results to the SFE object. MC stands for Monte Carlo. The
<code>nsim</code> argument specifies the number of simulations.</p>
<p>Add the results to the SFE object</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoranMC.html" class="external-link">colDataMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>, </span>
<span>                             colGraphName <span class="op">=</span> <span class="st">"visium"</span>, nsim <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/colFeatureData.html" class="external-link">colFeatureData</a></span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 10 rows and 10 columns</span></span>
<span><span class="co">#&gt;           MoransI_Vis5A   K_Vis5A GearysC_Vis5A MoranMC_statistic_Vis5A</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt; &lt;numeric&gt;     &lt;numeric&gt;               &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; barcode              NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; col                  NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; row                  NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; x                    NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; y                    NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; dia                  NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; tissue               NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; sample_id            NA        NA            NA                      NA</span></span>
<span><span class="co">#&gt; nCounts        0.528705   3.00082      0.474892                0.528705</span></span>
<span><span class="co">#&gt; nGenes         0.384028   3.88036      0.605797                0.384028</span></span>
<span><span class="co">#&gt;           MoranMC_parameter_Vis5A MoranMC_p.value_Vis5A</span></span>
<span><span class="co">#&gt;                         &lt;numeric&gt;             &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; barcode                        NA                    NA</span></span>
<span><span class="co">#&gt; col                            NA                    NA</span></span>
<span><span class="co">#&gt; row                            NA                    NA</span></span>
<span><span class="co">#&gt; x                              NA                    NA</span></span>
<span><span class="co">#&gt; y                              NA                    NA</span></span>
<span><span class="co">#&gt; dia                            NA                    NA</span></span>
<span><span class="co">#&gt; tissue                         NA                    NA</span></span>
<span><span class="co">#&gt; sample_id                      NA                    NA</span></span>
<span><span class="co">#&gt; nCounts                       101            0.00990099</span></span>
<span><span class="co">#&gt; nGenes                        101            0.00990099</span></span>
<span><span class="co">#&gt;           MoranMC_alternative_Vis5A   MoranMC_method_Vis5A</span></span>
<span><span class="co">#&gt;                         &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">#&gt; barcode                          NA                     NA</span></span>
<span><span class="co">#&gt; col                              NA                     NA</span></span>
<span><span class="co">#&gt; row                              NA                     NA</span></span>
<span><span class="co">#&gt; x                                NA                     NA</span></span>
<span><span class="co">#&gt; y                                NA                     NA</span></span>
<span><span class="co">#&gt; dia                              NA                     NA</span></span>
<span><span class="co">#&gt; tissue                           NA                     NA</span></span>
<span><span class="co">#&gt; sample_id                        NA                     NA</span></span>
<span><span class="co">#&gt; nCounts                     greater Monte-Carlo simulati..</span></span>
<span><span class="co">#&gt; nGenes                      greater Monte-Carlo simulati..</span></span>
<span><span class="co">#&gt;           MoranMC_data.name_Vis5A                          MoranMC_res_Vis5A</span></span>
<span><span class="co">#&gt;                       &lt;character&gt;                                     &lt;list&gt;</span></span>
<span><span class="co">#&gt; barcode                        NA                                         NA</span></span>
<span><span class="co">#&gt; col                            NA                                         NA</span></span>
<span><span class="co">#&gt; row                            NA                                         NA</span></span>
<span><span class="co">#&gt; x                              NA                                         NA</span></span>
<span><span class="co">#&gt; y                              NA                                         NA</span></span>
<span><span class="co">#&gt; dia                            NA                                         NA</span></span>
<span><span class="co">#&gt; tissue                         NA                                         NA</span></span>
<span><span class="co">#&gt; sample_id                      NA                                         NA</span></span>
<span><span class="co">#&gt; nCounts   x[i, ] \nweights: lis..  0.000970337,-0.010410652,-0.021168352,...</span></span>
<span><span class="co">#&gt; nGenes    x[i, ] \nweights: lis..          0.0242217,0.0348584,0.0274178,...</span></span></code></pre></div>
<p>Note that while the test is performed for multiple features, the
p-values here are not corrected for multiple hypothesis testing.</p>
<p>The results can be plotted:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotMoranMC.html" class="external-link">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-35-1.png" width="700" style="display: block; margin: auto;"></p>
<p>By default, the colorblind friendly palette from
<code>dittoSeq</code> is used for categorical variables. The density is
of Moran’s I from the simulations where the values are permuted and
disconnected from spatial locations, and the vertical line is the actual
Moran’s I value. The simulation indicates that the spatial
autocorrelation is significant.</p>
<p>Each function for Moran MC has a Geary’s C equivalent
(e.g. <code>runGearyMC</code>).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoranMC.html" class="external-link">colDataGearyMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>, </span>
<span>                             colGraphName <span class="op">=</span> <span class="st">"visium"</span>, nsim <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotMoranMC.html" class="external-link">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCounts"</span>, <span class="st">"nGenes"</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"GearyMC"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-37-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Also see some myofiber morphological metrics</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoranMC.html" class="external-link">annotGeometryMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, </span>
<span>                                   features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="st">"eccentricity"</span><span class="op">)</span>, </span>
<span>                                   annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span>,</span>
<span>                                   annotGraphName <span class="op">=</span> <span class="st">"myofiber_poly2nb"</span>, </span>
<span>                                   zero.policy <span class="op">=</span> <span class="cn">TRUE</span>, nsim <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>Specify <code>annotGeometryName</code> to make the
<code>plotMoranMC</code> function look for results in the
<code>annotGeometry</code> of interest.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotMoranMC.html" class="external-link">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="st">"eccentricity"</span><span class="op">)</span>,</span>
<span>            annotGeometryName <span class="op">=</span> <span class="st">"myofiber_simplified"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-39-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here all 4 morphological metrics have positive spatial
autocorrelation, but myofiber area is more strongly autocorrelated.</p>
<p>This can also be done to gene expression, again, by default, using
the “logcounts” assay.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoranMC.html" class="external-link">runMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, nsim <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>Although the row names of <code>sfe_tissue</code> are actually
Ensembl IDs, the human readable gene symbols are shown in the plots by
default.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotMoranMC.html" class="external-link">plotMoranMC</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, features <span class="op">=</span> <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-41-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The vertical lines for Car3 and Myh2 overlap but they’re not the same
value (see results of <code>runMoransI</code> above). These genes have
positive spatial autocorrelation that is stronger than total counts.</p>
</div>
<div class="section level4">
<h4 id="correlogram">Correlogram<a class="anchor" aria-label="anchor" href="#correlogram"></a>
</h4>
<p>What’s the length scale of the spatial autocorrelation? In a
correlogram, spatial autocorrelation of higher orders of neighbors
(e.g. second order neighbors are neighbors of neighbors) is calculated
to see how it decays over the orders. In Visium, with the regular
hexagonal grid, order of neighbors is a proxy for distance. For more
irregular patterns such as single cells, different methods to find the
spatial neighbors may give different results. Functions to compute
correlograms wrap <a href="https://r-spatial.github.io/spdep/reference/sp.correlogram.html" class="external-link"><code>sp.correlogram</code>
in <code>spdep</code></a>, and have the same pattern of
<code>calculate*</code> and <code>run*</code> as the Moran’s I and
permutation test functions, except for the <code>order</code> argument
specifying the maximum order of neighbors.</p>
<p>For <code>colData</code>, Moran’s I correlogram:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateCorrelogram.html" class="external-link">runCorrelogram</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span>, </span>
<span>                             order <span class="op">=</span> <span class="fl">10</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The results can be plotted with <code>plotCorrelogram</code></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotCorrelogram.html" class="external-link">plotCorrelogram</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="va">hvgs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-43-1.png" width="700" style="display: block; margin: auto;"></p>
<p>The error bars are twice the standard deviation of the Moran’s I
value. The standard deviation and p-values (null hypothesis is that
Moran’s I is 0) come from <code>moran.test</code> (for Geary’s C
correlogram, <code>geary.test</code>); these should be taken with a
grain of salt for data that is not normally distributed. The p-values
have been corrected for multiple hypothesis testing across all orders
and features. As usual, . means p &lt; 0.1, * means p &lt; 0.05, **
means p &lt; 0.01, and *** means p &lt; 0.001.</p>
<p>Again, this can be done for Geary’s C, <code>colData</code>,
<code>annotGeometry</code>, and etc. as in Moran’s I and permutation
testing.</p>
</div>
<div class="section level4">
<h4 id="moran-scatter-plot">Moran scatter plot<a class="anchor" aria-label="anchor" href="#moran-scatter-plot"></a>
</h4>
<p>In the Moran scatter plot, the x axis is the value itself and the y
axis is the average value of the neighbors. The slope of the fitted line
is Moran’s I. Sometimes clusters appear in this plot, showing different
kinds of neighborhoods for this value. Just like Moran’s I, permutation
testing, and correlogram functions, the functions for Moran scatter plot
also follow the <code>calculate*</code> and <code>run*</code> patterns
and have the same user interface. However, the plot can only be made for
one feature at a time.</p>
<p>For gene expression, to use one gene to demonstrate:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sfe_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/calculateMoranPlot.html" class="external-link">runMoranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh1"</span>, colGraphName <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/moranPlot.html" class="external-link">moranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh1"</span>, graphName <span class="op">=</span> <span class="st">"visium"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-45-1.png" width="700" style="display: block; margin: auto;"></p>
<p>There are no singletons here. Some Visium spots have neighbors that
don’t express Myh1 but spots that don’t express Myh1 usually have at
least some neighbors that do. There are 2 main clusters for spots whose
neighbors do express Myh1.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29</span><span class="op">)</span></span>
<span><span class="va">clusts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/clusterMoranPlot.html" class="external-link">clusterMoranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh1"</span>, BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/KmeansParam-class.html" class="external-link">KmeansParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/moranPlot.html" class="external-link">moranPlot</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh1"</span>, graphName <span class="op">=</span> <span class="st">"visium"</span>, color_by <span class="op">=</span> <span class="va">clusts</span><span class="op">$</span><span class="va">Myh1</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-47-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Again, plot the clusters in space</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">sfe_tissue</span><span class="op">)</span><span class="op">$</span><span class="va">Myh1_moranPlot_clust</span> <span class="op">&lt;-</span> <span class="va">clusts</span><span class="op">$</span><span class="va">Myh1</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Voyager/man/plotSpatialFeature.html" class="external-link">plotSpatialFeature</a></span><span class="op">(</span><span class="va">sfe_tissue</span>, <span class="st">"Myh1_moranPlot_clust"</span>, colGeometryName <span class="op">=</span> <span class="st">"spotPoly"</span><span class="op">)</span></span></code></pre></div>
<p><img src="vig3_Voyager_files/figure-html/unnamed-chunk-48-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Again, this can be done for <code>colData</code>,
<code>annotGeometry</code>, and etc. as in Moran’s I and permutation
testing.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 20.04.4 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] bluster_1.7.0                   BiocParallel_1.31.10           </span></span>
<span><span class="co">#&gt;  [3] sparseMatrixStats_1.9.0         patchwork_1.1.1                </span></span>
<span><span class="co">#&gt;  [5] scales_1.2.0                    sf_1.0-8                       </span></span>
<span><span class="co">#&gt;  [7] SFEData_0.99.0                  scran_1.25.0                   </span></span>
<span><span class="co">#&gt;  [9] scater_1.25.2                   ggplot2_3.3.6                  </span></span>
<span><span class="co">#&gt; [11] scuttle_1.7.2                   SingleCellExperiment_1.19.0    </span></span>
<span><span class="co">#&gt; [13] SummarizedExperiment_1.27.1     Biobase_2.57.1                 </span></span>
<span><span class="co">#&gt; [15] GenomicRanges_1.49.0            GenomeInfoDb_1.33.3            </span></span>
<span><span class="co">#&gt; [17] IRanges_2.31.0                  S4Vectors_0.35.1               </span></span>
<span><span class="co">#&gt; [19] BiocGenerics_0.43.0             MatrixGenerics_1.9.1           </span></span>
<span><span class="co">#&gt; [21] matrixStats_0.62.0              SpatialFeatureExperiment_0.99.0</span></span>
<span><span class="co">#&gt; [23] Voyager_0.99.0                 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] utf8_1.2.2                    R.utils_2.12.0               </span></span>
<span><span class="co">#&gt;   [3] tidyselect_1.1.2              RSQLite_2.2.15               </span></span>
<span><span class="co">#&gt;   [5] AnnotationDbi_1.59.1          grid_4.2.0                   </span></span>
<span><span class="co">#&gt;   [7] DropletUtils_1.17.0           munsell_0.5.0                </span></span>
<span><span class="co">#&gt;   [9] ScaledMatrix_1.5.0            codetools_0.2-18             </span></span>
<span><span class="co">#&gt;  [11] ragg_1.2.2                    units_0.8-0                  </span></span>
<span><span class="co">#&gt;  [13] statmod_1.4.36                withr_2.5.0                  </span></span>
<span><span class="co">#&gt;  [15] colorspace_2.0-3              filelock_1.0.2               </span></span>
<span><span class="co">#&gt;  [17] highr_0.9                     knitr_1.39                   </span></span>
<span><span class="co">#&gt;  [19] wk_0.6.0                      labeling_0.4.2               </span></span>
<span><span class="co">#&gt;  [21] GenomeInfoDbData_1.2.8        bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [23] farver_2.1.1                  rhdf5_2.41.1                 </span></span>
<span><span class="co">#&gt;  [25] rprojroot_2.0.3               vctrs_0.4.1                  </span></span>
<span><span class="co">#&gt;  [27] generics_0.1.3                xfun_0.31                    </span></span>
<span><span class="co">#&gt;  [29] BiocFileCache_2.5.0           R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [31] ggbeeswarm_0.6.0              rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt;  [33] locfit_1.5-9.6                isoband_0.2.5                </span></span>
<span><span class="co">#&gt;  [35] bitops_1.0-7                  rhdf5filters_1.9.0           </span></span>
<span><span class="co">#&gt;  [37] cachem_1.0.6                  DelayedArray_0.23.0          </span></span>
<span><span class="co">#&gt;  [39] assertthat_0.2.1              promises_1.2.0.1             </span></span>
<span><span class="co">#&gt;  [41] beeswarm_0.4.0                gtable_0.3.0                 </span></span>
<span><span class="co">#&gt;  [43] beachmat_2.13.4               rlang_1.0.4                  </span></span>
<span><span class="co">#&gt;  [45] systemfonts_1.0.4             splines_4.2.0                </span></span>
<span><span class="co">#&gt;  [47] scico_1.3.0                   BiocManager_1.30.18          </span></span>
<span><span class="co">#&gt;  [49] s2_1.1.0                      yaml_2.3.5                   </span></span>
<span><span class="co">#&gt;  [51] httpuv_1.6.5                  tools_4.2.0                  </span></span>
<span><span class="co">#&gt;  [53] spData_2.0.1                  SpatialExperiment_1.7.0      </span></span>
<span><span class="co">#&gt;  [55] ellipsis_0.3.2                raster_3.5-21                </span></span>
<span><span class="co">#&gt;  [57] jquerylib_0.1.4               RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;  [59] proxy_0.4-27                  Rcpp_1.0.9                   </span></span>
<span><span class="co">#&gt;  [61] zlibbioc_1.43.0               classInt_0.4-7               </span></span>
<span><span class="co">#&gt;  [63] purrr_0.3.4                   RCurl_1.98-1.7               </span></span>
<span><span class="co">#&gt;  [65] deldir_1.0-6                  viridis_0.6.2                </span></span>
<span><span class="co">#&gt;  [67] ggrepel_0.9.1                 cluster_2.1.3                </span></span>
<span><span class="co">#&gt;  [69] fs_1.5.2                      magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;  [71] magick_2.7.3                  ggnewscale_0.4.7             </span></span>
<span><span class="co">#&gt;  [73] mime_0.12                     evaluate_0.15                </span></span>
<span><span class="co">#&gt;  [75] xtable_1.8-4                  gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [77] compiler_4.2.0                tibble_3.1.7                 </span></span>
<span><span class="co">#&gt;  [79] KernSmooth_2.23-20            crayon_1.5.1                 </span></span>
<span><span class="co">#&gt;  [81] R.oo_1.25.0                   htmltools_0.5.3              </span></span>
<span><span class="co">#&gt;  [83] mgcv_1.8-40                   later_1.3.0                  </span></span>
<span><span class="co">#&gt;  [85] spdep_1.2-4                   DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [87] ExperimentHub_2.5.0           dbplyr_2.2.1                 </span></span>
<span><span class="co">#&gt;  [89] MASS_7.3-57                   rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [91] boot_1.3-28                   Matrix_1.4-1                 </span></span>
<span><span class="co">#&gt;  [93] cli_3.3.0                     R.methodsS3_1.8.2            </span></span>
<span><span class="co">#&gt;  [95] parallel_4.2.0                metapod_1.5.0                </span></span>
<span><span class="co">#&gt;  [97] igraph_1.3.4                  pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [99] pkgdown_2.0.6                 sp_1.5-0                     </span></span>
<span><span class="co">#&gt; [101] terra_1.5-34                  vipor_0.4.5                  </span></span>
<span><span class="co">#&gt; [103] bslib_0.4.0                   dqrng_0.3.0                  </span></span>
<span><span class="co">#&gt; [105] XVector_0.37.0                stringr_1.4.0                </span></span>
<span><span class="co">#&gt; [107] digest_0.6.29                 Biostrings_2.65.1            </span></span>
<span><span class="co">#&gt; [109] rmarkdown_2.14                edgeR_3.39.3                 </span></span>
<span><span class="co">#&gt; [111] DelayedMatrixStats_1.19.0     curl_4.3.2                   </span></span>
<span><span class="co">#&gt; [113] shiny_1.7.2                   rjson_0.2.21                 </span></span>
<span><span class="co">#&gt; [115] nlme_3.1-158                  lifecycle_1.0.1              </span></span>
<span><span class="co">#&gt; [117] jsonlite_1.8.0                Rhdf5lib_1.19.2              </span></span>
<span><span class="co">#&gt; [119] BiocNeighbors_1.15.1          desc_1.4.1                   </span></span>
<span><span class="co">#&gt; [121] viridisLite_0.4.0             limma_3.53.4                 </span></span>
<span><span class="co">#&gt; [123] fansi_1.0.3                   pillar_1.8.0                 </span></span>
<span><span class="co">#&gt; [125] lattice_0.20-45               KEGGREST_1.37.3              </span></span>
<span><span class="co">#&gt; [127] fastmap_1.1.0                 httr_1.4.3                   </span></span>
<span><span class="co">#&gt; [129] interactiveDisplayBase_1.35.0 glue_1.6.2                   </span></span>
<span><span class="co">#&gt; [131] png_0.1-7                     BiocVersion_3.16.0           </span></span>
<span><span class="co">#&gt; [133] bit_4.0.4                     class_7.3-20                 </span></span>
<span><span class="co">#&gt; [135] stringi_1.7.8                 sass_0.4.2                   </span></span>
<span><span class="co">#&gt; [137] HDF5Array_1.25.1              blob_1.2.3                   </span></span>
<span><span class="co">#&gt; [139] textshaping_0.3.6             BiocSingular_1.13.0          </span></span>
<span><span class="co">#&gt; [141] AnnotationHub_3.5.0           memoise_2.0.1                </span></span>
<span><span class="co">#&gt; [143] dplyr_1.0.9                   irlba_2.3.5                  </span></span>
<span><span class="co">#&gt; [145] e1071_1.7-11</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Lambda Moses, Lior Pachter.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
